<!DOCTYPE HTML>
<!-- NewPage -->
<html lang="de">
<head>
<!-- Generated by javadoc (15) on Wed Jul 20 11:19:49 CEST 2022 -->
<title>Bio</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="dc.created" content="2022-07-20">
<meta name="description" content="declaration: package: components, class: Bio">
<meta name="generator" content="javadoc/ClassWriterImpl">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
<script type="text/javascript" src="../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../script-dir/jquery-ui.min.js"></script>
</head>
<body class="class-declaration-page">
<script type="text/javascript">var data = {"i0":9,"i1":9,"i2":9,"i3":9,"i4":9,"i5":9,"i6":9,"i7":9,"i8":9,"i9":9};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],8:["t4","Concrete Methods"]};
var altColor = "alt-color";
var rowColor = "row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar.top">
<div class="skip-nav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.top.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li><a href="#nested.class.summary">Nested</a>&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search">SEARCH:</label>
<input type="text" id="search" value="search" disabled="disabled">
<input type="reset" id="reset" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip.navbar.top">
<!--   -->
</span></nav>
</header>
<div class="flex-content">
<main role="main">
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="sub-title"><span class="package-label-in-type">Package</span>&nbsp;<a href="package-summary.html">components</a></div>
<h1 title="Class Bio" class="title">Class Bio</h1>
</div>
<div class="inheritance" title="Inheritance Tree">java.lang.Object
<div class="inheritance">components.Bio</div>
</div>
<section class="description">
<hr>
<pre>public final class <span class="type-name-label">Bio</span>
extends java.lang.Object</pre>
<div class="block">Comprises static methods used to solve specifically bioinformatics problems.</div>
<dl class="notes">
<dt>Since:</dt>
<dd>2.0</dd>
<dt>Version:</dt>
<dd>2.1</dd>
<dt>Author:</dt>
<dd>Simon Hackl</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<!-- ======== NESTED CLASS SUMMARY ======== -->
<li>
<section class="nested-class-summary" id="nested.class.summary">
<h2>Nested Class Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Nested Classes</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Class</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<td class="col-first"><code>static class&nbsp;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a></span></code></th>
<td class="col-last">
<div class="block">Enum to store different modes to handle prefix gaps for global sequence alignment.</div>
</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- =========== FIELD SUMMARY =========== -->
<li>
<section class="field-summary" id="field.summary">
<h2>Field Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Fields</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Field</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.Character,&#8203;java.lang.Integer&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#AA1_PAM120_INDEX">AA1_PAM120_INDEX</a></span></code></th>
<td class="col-last">
<div class="block">Hash map mapping amino-acid one letter symbols to indices used to access substitution matrix values.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#AA1TO3">AA1TO3</a></span></code></th>
<td class="col-last">
<div class="block">Hash map mapping amino-acid one to three-letter codes.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#ANY_AA1">ANY_AA1</a></span></code></th>
<td class="col-last">
<div class="block">One-letter code character used to indicate the translation of an incomplete codon.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#ANY_AA3">ANY_AA3</a></span></code></th>
<td class="col-last">
<div class="block">Three-letter code string used to indicate the translation of an incomplete codon.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#CODON_MAP">CODON_MAP</a></span></code></th>
<td class="col-last">
<div class="block">Hash map mapping nucleotide codons to single-letter amino acids.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#DELETION_AA1">DELETION_AA1</a></span></code></th>
<td class="col-last">
<div class="block">One-letter code character used to indicate the deletion of an amino-acid.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#DELETION_AA3">DELETION_AA3</a></span></code></th>
<td class="col-last">
<div class="block">Three-letter code string used to indicate the deletion of an amino-acid.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#GAP">GAP</a></span></code></th>
<td class="col-last">
<div class="block">One-letter character used to indicate an alignment gap.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#NONE">NONE</a></span></code></th>
<td class="col-last">
<div class="block">Placeholder string to indicate no content.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>private static int[][]</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#PAM120">PAM120</a></span></code></th>
<td class="col-last">
<div class="block">The PAM120 matrix used for alignment computation.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#TERMINATION_AA1">TERMINATION_AA1</a></span></code></th>
<td class="col-last">
<div class="block">One-letter code character used to indicate a translated stop codon.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#TERMINATION_AA3">TERMINATION_AA3</a></span></code></th>
<td class="col-last">
<div class="block">Three-letter code String used to indicate a translated stop codon.</div>
</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<li>
<section class="constructor-summary" id="constructor.summary">
<h2>Constructor Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Constructors</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Constructor</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<th class="col-constructor-name" scope="row"><code><span class="member-name-link"><a href="#%3Cinit%3E()">Bio</a></span>()</code></th>
<td class="col-last">&nbsp;</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ========== METHOD SUMMARY =========== -->
<li>
<section class="method-summary" id="method.summary">
<h2>Method Summary</h2>
<div class="member-summary" id="method-summary-table">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button role="tab" aria-selected="true" aria-controls="method-summary-table.tabpanel" tabindex="0" onkeydown="switchTab(event)" id="t0" class="active-table-tab">All Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t1" class="table-tab" onclick="show(1);">Static Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t4" class="table-tab" onclick="show(8);">Concrete Methods</button></div>
<div id="method-summary-table.tabpanel" role="tabpanel">
<table class="summary-table" aria-labelledby="t0">
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Method</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color" id="i0">
<td class="col-first"><code>static java.util.ArrayList&lt;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getVariantsOfAlignedSequences(java.lang.String,java.lang.String)">getVariantsOfAlignedSequences</a></span>&#8203;(java.lang.String&nbsp;targetSequence,
java.lang.String&nbsp;querySequence)</code></th>
<td class="col-last">
<div class="block">Computes variants from two aligned sequences.</div>
</td>
</tr>
<tr class="row-color" id="i1">
<td class="col-first"><code>static org.javatuples.Triplet&lt;java.lang.Integer,&#8203;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#globalAminoAcidSequenceAlignment(java.lang.String,java.lang.String,int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)">globalAminoAcidSequenceAlignment</a></span>&#8203;(java.lang.String&nbsp;aaSeq1,
java.lang.String&nbsp;aaSeq2,
int&nbsp;gapOpenPenalty,
int&nbsp;gapExtendPenalty,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;left_mode,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;right_mode)</code></th>
<td class="col-last">
<div class="block">Invokes the <a href="#globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int%5B%5D%5B%5D,int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)"><code>globalSequenceAlignment(java.lang.String, java.lang.String, java.util.HashMap&lt;java.lang.Character, java.lang.Integer&gt;, int[][], int, int, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)</code></a> method with pre-specified parameters for amino-acid sequence
 alignment.</div>
</td>
</tr>
<tr class="alt-color" id="i2">
<td class="col-first"><code>static org.javatuples.Triplet&lt;java.lang.Integer,&#8203;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#globalNucleotideSequenceAlignment(java.lang.String,java.lang.String,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)">globalNucleotideSequenceAlignment</a></span>&#8203;(java.lang.String&nbsp;nucSeq1,
java.lang.String&nbsp;nucSeq2,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;left_mode,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;right_mode)</code></th>
<td class="col-last">
<div class="block">Invokes the <a href="#globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int%5B%5D%5B%5D,int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)"><code>globalSequenceAlignment(java.lang.String, java.lang.String, java.util.HashMap&lt;java.lang.Character, java.lang.Integer&gt;, int[][], int, int, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)</code></a> method with pre-specified parameters for nucleotide sequence
 alignment.</div>
</td>
</tr>
<tr class="row-color" id="i3">
<td class="col-first"><code>static org.javatuples.Triplet&lt;java.lang.Integer,&#8203;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int%5B%5D%5B%5D,int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)">globalSequenceAlignment</a></span>&#8203;(java.lang.String&nbsp;seq1,
java.lang.String&nbsp;seq2,
java.util.HashMap&lt;java.lang.Character,&#8203;java.lang.Integer&gt;&nbsp;scoringMatrixIndexMap,
int[][]&nbsp;scoringMatrix,
int&nbsp;gapOpenPenalty,
int&nbsp;gapExtendPenalty,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;left_mode,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;right_mode)</code></th>
<td class="col-last">
<div class="block">Computes a global sequence alignment using the gap-affine Needleman-Wunsch algorithm.</div>
</td>
</tr>
<tr class="alt-color" id="i4">
<td class="col-first"><code>static java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#inferProteoform(datastructure.VariantsDictionary,java.lang.String,java.lang.String)">inferProteoform</a></span>&#8203;(<a href="../datastructure/VariantsDictionary.html" title="class in datastructure">VariantsDictionary</a>&nbsp;variantsDictionary,
java.lang.String&nbsp;fId,
java.lang.String&nbsp;sId)</code></th>
<td class="col-last">
<div class="block">Infers the proteoform variants of a single sample wrt. a single feature from the information of the passed
 variantsDictionary.</div>
</td>
</tr>
<tr class="row-color" id="i5">
<td class="col-first"><code>static java.lang.Character</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#invertBase(char)">invertBase</a></span>&#8203;(char&nbsp;base)</code></th>
<td class="col-last">
<div class="block">Returns the complement of a nucleotide base, i.e. switches A to T, C to G and vice versa.</div>
</td>
</tr>
<tr class="alt-color" id="i6">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#reverseComplement(java.lang.String)">reverseComplement</a></span>&#8203;(java.lang.String&nbsp;sequence)</code></th>
<td class="col-last">
<div class="block">Returns the reverse complement of the passed nucleotide sequence.</div>
</td>
</tr>
<tr class="row-color" id="i7">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#translateCodon(java.lang.String,boolean,boolean,boolean)">translateCodon</a></span>&#8203;(java.lang.String&nbsp;codon,
boolean&nbsp;asAA3,
boolean&nbsp;includeTermination,
boolean&nbsp;includeIncomplete)</code></th>
<td class="col-last">
<div class="block">Returns an amino-acid representation of the passed `codon`.</div>
</td>
</tr>
<tr class="alt-color" id="i8">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#translateNucSequence(java.lang.Iterable,boolean,boolean)">translateNucSequence</a></span>&#8203;(java.lang.Iterable&lt;java.lang.String&gt;&nbsp;splitNucSequence,
boolean&nbsp;includeTermination,
boolean&nbsp;includeIncomplete)</code></th>
<td class="col-last">
<div class="block">Translates a nucleotide sequence, split into codons of length three, into a single-letter amino acid sequence.</div>
</td>
</tr>
<tr class="row-color" id="i9">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#translateNucSequence(java.lang.String,boolean,boolean,boolean)">translateNucSequence</a></span>&#8203;(java.lang.String&nbsp;nucSequence,
boolean&nbsp;includeTermination,
boolean&nbsp;includeIncomplete,
boolean&nbsp;asSense)</code></th>
<td class="col-last">
<div class="block">Translates a nucleotide sequence into a single-letter amino acid sequence.</div>
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="inherited-list">
<h3 id="methods.inherited.from.class.java.lang.Object">Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></div>
</section>
</li>
</ul>
</section>
<section class="details">
<ul class="details-list">
<!-- ============ FIELD DETAIL =========== -->
<li>
<section class="field-details" id="field.detail">
<h2>Field Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="TERMINATION_AA1">
<h3>TERMINATION_AA1</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">TERMINATION_AA1</span></div>
<div class="block">One-letter code character used to indicate a translated stop codon.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#components.Bio.TERMINATION_AA1">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="TERMINATION_AA3">
<h3>TERMINATION_AA3</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">TERMINATION_AA3</span></div>
<div class="block">Three-letter code String used to indicate a translated stop codon.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#components.Bio.TERMINATION_AA3">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="ANY_AA1">
<h3>ANY_AA1</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">ANY_AA1</span></div>
<div class="block">One-letter code character used to indicate the translation of an incomplete codon.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#components.Bio.ANY_AA1">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="ANY_AA3">
<h3>ANY_AA3</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">ANY_AA3</span></div>
<div class="block">Three-letter code string used to indicate the translation of an incomplete codon.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#components.Bio.ANY_AA3">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="DELETION_AA1">
<h3>DELETION_AA1</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">DELETION_AA1</span></div>
<div class="block">One-letter code character used to indicate the deletion of an amino-acid.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#components.Bio.DELETION_AA1">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="DELETION_AA3">
<h3>DELETION_AA3</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">DELETION_AA3</span></div>
<div class="block">Three-letter code string used to indicate the deletion of an amino-acid.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#components.Bio.DELETION_AA3">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="GAP">
<h3>GAP</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">GAP</span></div>
<div class="block">One-letter character used to indicate an alignment gap.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#components.Bio.GAP">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="NONE">
<h3>NONE</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">NONE</span></div>
<div class="block">Placeholder string to indicate no content.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#components.Bio.NONE">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="CODON_MAP">
<h3>CODON_MAP</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">CODON_MAP</span></div>
<div class="block">Hash map mapping nucleotide codons to single-letter amino acids. Stop codons map to no character.</div>
</section>
</li>
<li>
<section class="detail" id="AA1TO3">
<h3>AA1TO3</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">AA1TO3</span></div>
<div class="block">Hash map mapping amino-acid one to three-letter codes.</div>
</section>
</li>
<li>
<section class="detail" id="AA1_PAM120_INDEX">
<h3>AA1_PAM120_INDEX</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.Character,&#8203;java.lang.Integer&gt;</span>&nbsp;<span class="member-name">AA1_PAM120_INDEX</span></div>
<div class="block">Hash map mapping amino-acid one letter symbols to indices used to access substitution matrix values.</div>
</section>
</li>
<li>
<section class="detail" id="PAM120">
<h3>PAM120</h3>
<div class="member-signature"><span class="modifiers">private static final</span>&nbsp;<span class="return-type">int[][]</span>&nbsp;<span class="member-name">PAM120</span></div>
<div class="block">The PAM120 matrix used for alignment computation.
 <p>
 - Computed with <a href="https://bioinformaticshome.com/online_software/make_pam/makePAMmatrix.html">makePAMmatrix.html</a>
 and a scaling factor of 1.
 - Includes values for 'X'/any amino-acid (match/mismatch: -1) and '*'/termination (match/mismatch: -4/-1).</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<li>
<section class="constructor-details" id="constructor.detail">
<h2>Constructor Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="&lt;init&gt;()">
<h3>Bio</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="member-name">Bio</span>()</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ============ METHOD DETAIL ========== -->
<li>
<section class="method-details" id="method.detail">
<h2>Method Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="translateCodon(java.lang.String,boolean,boolean,boolean)">
<h3>translateCodon</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">translateCodon</span>&#8203;(<span class="parameters">java.lang.String&nbsp;codon,
boolean&nbsp;asAA3,
boolean&nbsp;includeTermination,
boolean&nbsp;includeIncomplete)</span>
                                       throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Returns an amino-acid representation of the passed `codon`.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>codon</code> - <code>String</code> representing the nucleotide codon to translate.</dd>
<dd><code>asAA3</code> - <code>Boolean</code> whether the codon should be translated to the amino-acid three-letter code or not.</dd>
<dd><code>includeTermination</code> - <code>Boolean</code> whether a representation for translation termination (in the case of
                           three-letter code 'TER' and in the case of one-letter code 'Z') shall
                           be returned or not, i.e. an empty String is returned instead.</dd>
<dd><code>includeIncomplete</code> - <code>Boolean</code> whether incomplete codons, i.e. with a length other than 3, should be
                           translated as incomplete amino-acids.</dd>
<dt>Returns:</dt>
<dd><code>String</code> representing a translated amino-acid.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If the passed codon has a length other than three and `includeIncomplete` is false.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="translateNucSequence(java.lang.Iterable,boolean,boolean)">
<h3>translateNucSequence</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">translateNucSequence</span>&#8203;(<span class="parameters">java.lang.Iterable&lt;java.lang.String&gt;&nbsp;splitNucSequence,
boolean&nbsp;includeTermination,
boolean&nbsp;includeIncomplete)</span>
                                             throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Translates a nucleotide sequence, split into codons of length three, into a single-letter amino acid sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>splitNucSequence</code> - <code>java.util.ArrayList&lt;java.lang.String&gt;</code> representing a nucleotide sequence that was split into codons
                           of length 3.</dd>
<dd><code>includeTermination</code> - <code>Boolean</code> whether a representation for translation termination (in the case of
                           three-letter code 'TER' and in the case of one-letter code 'Z') shall
                           be returned or not, i.e. an empty String is returned instead.</dd>
<dd><code>includeIncomplete</code> - <code>Boolean</code> whether an incomplete amino-acid should be added to the end if the
                           sequence contains an incomplete codon at the end.</dd>
<dt>Returns:</dt>
<dd><code>String</code> representing the translated nucleotide sequence.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If any codon with a length different than three is detected.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="translateNucSequence(java.lang.String,boolean,boolean,boolean)">
<h3>translateNucSequence</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">translateNucSequence</span>&#8203;(<span class="parameters">java.lang.String&nbsp;nucSequence,
boolean&nbsp;includeTermination,
boolean&nbsp;includeIncomplete,
boolean&nbsp;asSense)</span>
                                             throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Translates a nucleotide sequence into a single-letter amino acid sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>nucSequence</code> - <code>String</code> representing a nucleotide sequence.</dd>
<dd><code>includeTermination</code> - <code>Boolean</code> whether a representation for translation termination (in the case of
                           three-letter code 'TER' and in the case of one-letter code 'Z') shall
                           be returned or not, i.e. an empty String is returned instead.</dd>
<dd><code>includeIncomplete</code> - <code>Boolean</code> whether an incomplete amino-acid should be added to the end if the
                           sequence contains an incomplete codon at the end.</dd>
<dd><code>asSense</code> - <code>Boolean</code> whether the sequence shall be translated as sense or anti-sense.</dd>
<dt>Returns:</dt>
<dd><code>String</code> representing the translated nucleotide sequence.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If any codon with a length different than three is detected.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="reverseComplement(java.lang.String)">
<h3>reverseComplement</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">reverseComplement</span>&#8203;(<span class="parameters">java.lang.String&nbsp;sequence)</span></div>
<div class="block">Returns the reverse complement of the passed nucleotide sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>sequence</code> - <code>String</code> representing a nucleotide sequence.</dd>
<dt>Returns:</dt>
<dd><code>String</code> representing the reverse complement of the passed sequence.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="invertBase(char)">
<h3>invertBase</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.Character</span>&nbsp;<span class="member-name">invertBase</span>&#8203;(<span class="parameters">char&nbsp;base)</span></div>
<div class="block">Returns the complement of a nucleotide base, i.e. switches A to T, C to G and vice versa. For non nucleotide
 symbols the identity is returned.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>base</code> - <code>Character</code> base to invert.</dd>
<dt>Returns:</dt>
<dd><code>Character</code> the inverted base.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="globalNucleotideSequenceAlignment(java.lang.String,java.lang.String,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)">
<h3>globalNucleotideSequenceAlignment</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">org.javatuples.Triplet&lt;java.lang.Integer,&#8203;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">globalNucleotideSequenceAlignment</span>&#8203;(<span class="parameters">java.lang.String&nbsp;nucSeq1,
java.lang.String&nbsp;nucSeq2,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;left_mode,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;right_mode)</span></div>
<div class="block">Invokes the <a href="#globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int%5B%5D%5B%5D,int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)"><code>globalSequenceAlignment(java.lang.String, java.lang.String, java.util.HashMap&lt;java.lang.Character, java.lang.Integer&gt;, int[][], int, int, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)</code></a> method with pre-specified parameters for nucleotide sequence
 alignment.
 <p>
 - Uses a simple substitution matrix that scores matches with 1 and mismatches with -1.
 - Uses a gap open and extension penalty of -2 and -1, respectively.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>nucSeq1</code> - <code>String</code> representation of the first nucleotide sequence for alignment.</dd>
<dd><code>nucSeq2</code> - <code>String</code> representation of the second nucleotide sequence for alignment.</dd>
<dd><code>left_mode</code> - <a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components"><code>Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</code></a> value to indicate how to handle left-marginal gaps.</dd>
<dd><code>right_mode</code> - <a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components"><code>Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</code></a> value to indicate how to handle right-marginal gaps.</dd>
<dt>Returns:</dt>
<dd><code>Triplet</code> storing the alignment score, the aligned first sequence and the aligned second sequence.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="globalAminoAcidSequenceAlignment(java.lang.String,java.lang.String,int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)">
<h3>globalAminoAcidSequenceAlignment</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">org.javatuples.Triplet&lt;java.lang.Integer,&#8203;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">globalAminoAcidSequenceAlignment</span>&#8203;(<span class="parameters">java.lang.String&nbsp;aaSeq1,
java.lang.String&nbsp;aaSeq2,
int&nbsp;gapOpenPenalty,
int&nbsp;gapExtendPenalty,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;left_mode,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;right_mode)</span></div>
<div class="block">Invokes the <a href="#globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int%5B%5D%5B%5D,int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)"><code>globalSequenceAlignment(java.lang.String, java.lang.String, java.util.HashMap&lt;java.lang.Character, java.lang.Integer&gt;, int[][], int, int, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)</code></a> method with pre-specified parameters for amino-acid sequence
 alignment.
 <p>
 - Uses the PAM120 substitution matrix.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>aaSeq1</code> - <code>String</code> representation of the first amino-acid sequence for alignment.</dd>
<dd><code>aaSeq2</code> - <code>String</code> representation of the second amino-acid sequence for alignment.</dd>
<dd><code>gapOpenPenalty</code> - <code>Integer</code> (positive!) to use as gap open penalty.</dd>
<dd><code>gapExtendPenalty</code> - <code>Integer</code> (positive!) to use as gap extension penalty.</dd>
<dd><code>left_mode</code> - <a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components"><code>Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</code></a> value to indicate how to handle left-marginal gaps.</dd>
<dd><code>right_mode</code> - <a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components"><code>Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</code></a> value to indicate how to handle right-marginal gaps.</dd>
<dt>Returns:</dt>
<dd><code>Triplet</code> storing the alignment score, the aligned first sequence and the aligned second sequence.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int[][],int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)">
<h3>globalSequenceAlignment</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">org.javatuples.Triplet&lt;java.lang.Integer,&#8203;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">globalSequenceAlignment</span>&#8203;(<span class="parameters">java.lang.String&nbsp;seq1,
java.lang.String&nbsp;seq2,
java.util.HashMap&lt;java.lang.Character,&#8203;java.lang.Integer&gt;&nbsp;scoringMatrixIndexMap,
int[][]&nbsp;scoringMatrix,
int&nbsp;gapOpenPenalty,
int&nbsp;gapExtendPenalty,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;left_mode,
<a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a>&nbsp;right_mode)</span></div>
<div class="block">Computes a global sequence alignment using the gap-affine Needleman-Wunsch algorithm.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>seq1</code> - <code>String</code> representation of the first sequence for alignment.</dd>
<dd><code>seq2</code> - <code>String</code> representation of the second sequence for alignment.</dd>
<dd><code>scoringMatrixIndexMap</code> - <code>HashMap</code> mapping characters that may occur in the aligned sequences to integer indices of the scoring matrix.</dd>
<dd><code>scoringMatrix</code> - <code>Integer</code> containing the scoring values for each pair of characters that may occur in the aligned sequences.</dd>
<dd><code>gapOpenPenalty</code> - <code>Integer</code> value used to penalize the opening of a gap.</dd>
<dd><code>gapExtendPenalty</code> - <code>Integer</code> value used to penalize the extension of a gap.</dd>
<dd><code>left_mode</code> - <a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components"><code>Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</code></a> value to indicate how to handle left-marginal gaps.</dd>
<dd><code>right_mode</code> - <a href="Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components"><code>Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</code></a> value to indicate how to handle right-marginal gaps.</dd>
<dt>Returns:</dt>
<dd><code>Triplet</code> storing the alignment score, the aligned first sequence and the aligned second sequence.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getVariantsOfAlignedSequences(java.lang.String,java.lang.String)">
<h3>getVariantsOfAlignedSequences</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.ArrayList&lt;java.lang.String&gt;</span>&nbsp;<span class="member-name">getVariantsOfAlignedSequences</span>&#8203;(<span class="parameters">java.lang.String&nbsp;targetSequence,
java.lang.String&nbsp;querySequence)</span></div>
<div class="block">Computes variants from two aligned sequences.
 <p>
 Variants are extracted wrt. the query versus the target sequence and formatted as p@r@a, where p is the variant
 position, r is the reference content and a is the single alternate alleles content..</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>targetSequence</code> - <code>String</code> representation of the target sequence (i.e. reference).</dd>
<dd><code>querySequence</code> - <code>String</code> representation of the query sequence (i.e. the one with variants).</dd>
<dt>Returns:</dt>
<dd><code>ArrayList</code> containing derived variants, c.f. method description for format details.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="inferProteoform(datastructure.VariantsDictionary,java.lang.String,java.lang.String)">
<h3>inferProteoform</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">inferProteoform</span>&#8203;(<span class="parameters"><a href="../datastructure/VariantsDictionary.html" title="class in datastructure">VariantsDictionary</a>&nbsp;variantsDictionary,
java.lang.String&nbsp;fId,
java.lang.String&nbsp;sId)</span>
                                                                                                     throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Infers the proteoform variants of a single sample wrt. a single feature from the information of the passed
 variantsDictionary.
 <p>
 - The computed variants maps keys are formatted as `x+y`, where x is the 1-based indexed position in the
 reference sequence at which the variant occurs and y is the 1-based indexed number of inserted positions after
 this position.
 - The computed variants maps values are the respective single letter code amino-acid contents of the variants
 including the content at the variant position.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>variantsDictionary</code> - <a href="../datastructure/VariantsDictionary.html" title="class in datastructure"><code>VariantsDictionary</code></a> instance which content is used to infer proteoform information.</dd>
<dd><code>fId</code> - <code>String</code> specifying the <a href="../datastructure/FeatureEntry.html#name"><code>FeatureEntry.name</code></a> of the feature to which respect
                           the sample proteoform should be inferred.</dd>
<dd><code>sId</code> - <code>String</code> specifying the <a href="../datastructure/SampleEntry.html#name"><code>SampleEntry.name</code></a> of the sample of
                           which the proteoform should be inferred.</dd>
<dt>Returns:</dt>
<dd><code>ConcurrentSkipListMap</code> mapping positions to variant contents.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If any translation procedure of nucleotide sequences fails.</dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
</ul>
</section>
<!-- ========= END OF CLASS DATA ========= -->
</main>
<footer role="contentinfo">
<nav role="navigation">
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottom-nav" id="navbar.bottom">
<div class="skip-nav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.bottom.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li><a href="#nested.class.summary">Nested</a>&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
</div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<span class="skip-nav" id="skip.navbar.bottom">
<!--   -->
</span></nav>
</footer>
</div>
</div>
</body>
</html>
