<!DOCTYPE HTML>
<!-- NewPage -->
<html lang="de">
<head>
<!-- Generated by javadoc (15) on Mon Jul 25 13:17:44 CEST 2022 -->
<title>IO</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="dc.created" content="2022-07-25">
<meta name="description" content="declaration: package: utility, class: IO">
<meta name="generator" content="javadoc/ClassWriterImpl">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
<script type="text/javascript" src="../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../script-dir/jquery-ui.min.js"></script>
</head>
<body class="class-declaration-page">
<script type="text/javascript">var data = {"i0":9,"i1":9,"i2":9,"i3":9,"i4":9,"i5":9,"i6":9,"i7":9,"i8":9,"i9":9,"i10":9,"i11":9};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],8:["t4","Concrete Methods"]};
var altColor = "alt-color";
var rowColor = "row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar.top">
<div class="skip-nav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.top.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search">SEARCH:</label>
<input type="text" id="search" value="search" disabled="disabled">
<input type="reset" id="reset" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip.navbar.top">
<!--   -->
</span></nav>
</header>
<div class="flex-content">
<main role="main">
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="sub-title"><span class="package-label-in-type">Package</span>&nbsp;<a href="package-summary.html">utility</a></div>
<h1 title="Class IO" class="title">Class IO</h1>
</div>
<div class="inheritance" title="Inheritance Tree">java.lang.Object
<div class="inheritance">utility.IO</div>
</div>
<section class="description">
<hr>
<pre>public final class <span class="type-name-label">IO</span>
extends java.lang.Object</pre>
<div class="block">This class comprises static methods used for reading and writing files.</div>
<dl class="notes">
<dt>Since:</dt>
<dd>2.0</dd>
<dt>Version:</dt>
<dd>2.1</dd>
<dt>Author:</dt>
<dd>Simon Hackl</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<!-- =========== FIELD SUMMARY =========== -->
<li>
<section class="field-summary" id="field.summary">
<h2>Field Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Fields</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Field</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#LINE_SEPARATOR">LINE_SEPARATOR</a></span></code></th>
<td class="col-last">
<div class="block">OS dependent line separator</div>
</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<li>
<section class="constructor-summary" id="constructor.summary">
<h2>Constructor Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Constructors</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Constructor</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<th class="col-constructor-name" scope="row"><code><span class="member-name-link"><a href="#%3Cinit%3E()">IO</a></span>()</code></th>
<td class="col-last">&nbsp;</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ========== METHOD SUMMARY =========== -->
<li>
<section class="method-summary" id="method.summary">
<h2>Method Summary</h2>
<div class="member-summary" id="method-summary-table">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button role="tab" aria-selected="true" aria-controls="method-summary-table.tabpanel" tabindex="0" onkeydown="switchTab(event)" id="t0" class="active-table-tab">All Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t1" class="table-tab" onclick="show(1);">Static Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t4" class="table-tab" onclick="show(8);">Concrete Methods</button></div>
<div id="method-summary-table.tabpanel" role="tabpanel">
<table class="summary-table" aria-labelledby="t0">
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Method</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color" id="i0">
<td class="col-first"><code>static void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#copyFile(java.io.File,java.io.File)">copyFile</a></span>&#8203;(java.io.File&nbsp;file,
java.io.File&nbsp;target)</code></th>
<td class="col-last">
<div class="block">Copies the file pointed to with file to the file pointed to with target.</div>
</td>
</tr>
<tr class="row-color" id="i1">
<td class="col-first"><code>static void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#deleteDirectory(java.io.File)">deleteDirectory</a></span>&#8203;(java.io.File&nbsp;file)</code></th>
<td class="col-last">
<div class="block">Accepts a <code>File</code> object representing a directory that is subject to deletion.</div>
</td>
</tr>
<tr class="alt-color" id="i2">
<td class="col-first"><code>static void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#generateDirectory(java.io.File)">generateDirectory</a></span>&#8203;(java.io.File&nbsp;file)</code></th>
<td class="col-last">
<div class="block">Tries to generate the directory specified by the passed <code>File</code> object.</div>
</td>
</tr>
<tr class="row-color" id="i3">
<td class="col-first"><code>static void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#generateFile(java.io.File)">generateFile</a></span>&#8203;(java.io.File&nbsp;file)</code></th>
<td class="col-last">
<div class="block">Tries to generate the file specified by the passed <code>File</code> object.</div>
</td>
</tr>
<tr class="alt-color" id="i4">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getSequencesFromPdbStructure(org.biojava.nbio.structure.Structure)">getSequencesFromPdbStructure</a></span>&#8203;(org.biojava.nbio.structure.Structure&nbsp;structure)</code></th>
<td class="col-last">
<div class="block">Returns the chains sequences from a <code>Structure</code> instance parsed from a .pdb file.</div>
</td>
</tr>
<tr class="row-color" id="i5">
<td class="col-first"><code>static java.util.HashSet&lt;<a href="../datastructure/FastaContainer.html" title="class in datastructure">FastaContainer</a>&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#readFastaToSet(java.io.File)">readFastaToSet</a></span>&#8203;(java.io.File&nbsp;file)</code></th>
<td class="col-last">
<div class="block">Reads a .fasta file from a <code>File</code> object into a <code>HashSet</code> of <a href="../datastructure/FastaContainer.html" title="class in datastructure"><code>FastaContainer</code></a> instances.</div>
</td>
</tr>
<tr class="alt-color" id="i6">
<td class="col-first"><code>static org.biojava.nbio.genome.parsers.gff.FeatureList</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#readGFF(java.io.File)">readGFF</a></span>&#8203;(java.io.File&nbsp;file)</code></th>
<td class="col-last">
<div class="block">Reads a .gff file from the path specified by a <code>File</code> object and returns a
 <code>FeatureList</code> object.</div>
</td>
</tr>
<tr class="row-color" id="i7">
<td class="col-first"><code>static java.util.ArrayList&lt;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#readLinesFromFile(java.lang.String)">readLinesFromFile</a></span>&#8203;(java.lang.String&nbsp;filePath)</code></th>
<td class="col-last">
<div class="block">Reads a file and returns its content line-wise as list.</div>
</td>
</tr>
<tr class="alt-color" id="i8">
<td class="col-first"><code>static org.biojava.nbio.structure.Structure</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#readStructure(java.io.File)">readStructure</a></span>&#8203;(java.io.File&nbsp;pdbFile)</code></th>
<td class="col-last">
<div class="block">Parses a <code>Structure</code> instance from a <code>File</code> pointing to a pdb file.</div>
</td>
</tr>
<tr class="row-color" id="i9">
<td class="col-first"><code>static <a href="../datastructure/VariantsDictionary.html" title="class in datastructure">VariantsDictionary</a></code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#readVariantsDictionary(java.io.File)">readVariantsDictionary</a></span>&#8203;(java.io.File&nbsp;vDictFile)</code></th>
<td class="col-last">
<div class="block">Initializes a <a href="../datastructure/VariantsDictionary.html" title="class in datastructure"><code>VariantsDictionary</code></a> object from a variants dictionary JSON file.</div>
</td>
</tr>
<tr class="alt-color" id="i10">
<td class="col-first"><code>static void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#writeFasta(java.io.File,java.util.HashMap)">writeFasta</a></span>&#8203;(java.io.File&nbsp;outputFile,
java.util.HashMap&lt;java.lang.String,&#8203;java.util.ArrayList&lt;java.lang.String&gt;&gt;&nbsp;sequences)</code></th>
<td class="col-last">
<div class="block">Writes a fasta format file to the specified output file from a <code>HashMap</code> instance mapping sequences to
 lists of identifiers (used to construct the fasta entry headers).</div>
</td>
</tr>
<tr class="row-color" id="i11">
<td class="col-first"><code>static void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#writeVcf(java.io.File,java.util.NavigableSet,java.lang.String)">writeVcf</a></span>&#8203;(java.io.File&nbsp;outputFile,
java.util.NavigableSet&lt;java.lang.String&gt;&nbsp;variants,
java.lang.String&nbsp;chrom)</code></th>
<td class="col-last">
<div class="block">Writes a dummy vcf format file to the specified output file from a <code>NavigableSet</code> instance yielding
 <code>String</code>s that represent variants.</div>
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="inherited-list">
<h3 id="methods.inherited.from.class.java.lang.Object">Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></div>
</section>
</li>
</ul>
</section>
<section class="details">
<ul class="details-list">
<!-- ============ FIELD DETAIL =========== -->
<li>
<section class="field-details" id="field.detail">
<h2>Field Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="LINE_SEPARATOR">
<h3>LINE_SEPARATOR</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">LINE_SEPARATOR</span></div>
<div class="block">OS dependent line separator</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<li>
<section class="constructor-details" id="constructor.detail">
<h2>Constructor Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="&lt;init&gt;()">
<h3>IO</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="member-name">IO</span>()</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ============ METHOD DETAIL ========== -->
<li>
<section class="method-details" id="method.detail">
<h2>Method Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="readLinesFromFile(java.lang.String)">
<h3>readLinesFromFile</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.ArrayList&lt;java.lang.String&gt;</span>&nbsp;<span class="member-name">readLinesFromFile</span>&#8203;(<span class="parameters">java.lang.String&nbsp;filePath)</span>
                                                               throws <span class="exceptions">java.io.FileNotFoundException</span></div>
<div class="block">Reads a file and returns its content line-wise as list.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>filePath</code> - <code>String</code> representing a file path.</dd>
<dt>Returns:</dt>
<dd><code>ArrayList</code> of <code>String</code> objects, each representing one line of the file accessed via the
 passed file path.</dd>
<dt>Throws:</dt>
<dd><code>java.io.FileNotFoundException</code> - If the specified file path does not lead to any file.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="readGFF(java.io.File)">
<h3>readGFF</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">org.biojava.nbio.genome.parsers.gff.FeatureList</span>&nbsp;<span class="member-name">readGFF</span>&#8203;(<span class="parameters">java.io.File&nbsp;file)</span>
                                                               throws <span class="exceptions">java.io.IOException</span></div>
<div class="block">Reads a .gff file from the path specified by a <code>File</code> object and returns a
 <code>FeatureList</code> object.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>file</code> - <code>File</code> specifying a .gff file.</dd>
<dt>Returns:</dt>
<dd><code>FeatureList</code> comprising the entries from the read .gff file.</dd>
<dt>Throws:</dt>
<dd><code>java.io.IOException</code> - If the <code>GFF3Reader</code> throws any <code>IOException</code>.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="readFastaToSet(java.io.File)">
<h3>readFastaToSet</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.HashSet&lt;<a href="../datastructure/FastaContainer.html" title="class in datastructure">FastaContainer</a>&gt;</span>&nbsp;<span class="member-name">readFastaToSet</span>&#8203;(<span class="parameters">java.io.File&nbsp;file)</span>
                                                        throws <span class="exceptions">java.io.IOException</span></div>
<div class="block">Reads a .fasta file from a <code>File</code> object into a <code>HashSet</code> of <a href="../datastructure/FastaContainer.html" title="class in datastructure"><code>FastaContainer</code></a> instances. Each
 element contains the header and sequence of the respective entry from the .fasta file.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>file</code> - <code>File</code> specifying a .fasta file.</dd>
<dt>Returns:</dt>
<dd><code>HashSet</code> of <a href="../datastructure/FastaContainer.html" title="class in datastructure"><code>FastaContainer</code></a> instances.</dd>
<dt>Throws:</dt>
<dd><code>java.io.IOException</code> - If the respective file can not be found or accessed.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="readVariantsDictionary(java.io.File)">
<h3>readVariantsDictionary</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type"><a href="../datastructure/VariantsDictionary.html" title="class in datastructure">VariantsDictionary</a></span>&nbsp;<span class="member-name">readVariantsDictionary</span>&#8203;(<span class="parameters">java.io.File&nbsp;vDictFile)</span>
                                                 throws <span class="exceptions">java.io.IOException</span></div>
<div class="block">Initializes a <a href="../datastructure/VariantsDictionary.html" title="class in datastructure"><code>VariantsDictionary</code></a> object from a variants dictionary JSON file.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>vDictFile</code> - <code>File</code> pointing to the JSON file to parse.</dd>
<dt>Returns:</dt>
<dd><a href="../datastructure/VariantsDictionary.html" title="class in datastructure"><code>VariantsDictionary</code></a> instance.</dd>
<dt>Throws:</dt>
<dd><code>java.io.IOException</code> - If any error occurs while parsing the variants dictionary JSON file.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="deleteDirectory(java.io.File)">
<h3>deleteDirectory</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">deleteDirectory</span>&#8203;(<span class="parameters">java.io.File&nbsp;file)</span>
                            throws <span class="exceptions">java.io.IOException</span></div>
<div class="block">Accepts a <code>File</code> object representing a directory that is subject to deletion.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>file</code> - <code>File</code> specifying the directory to delete.</dd>
<dt>Throws:</dt>
<dd><code>java.io.IOException</code> - If any file deletion procedure failed, for example the specified directory does not exist.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="generateDirectory(java.io.File)">
<h3>generateDirectory</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">generateDirectory</span>&#8203;(<span class="parameters">java.io.File&nbsp;file)</span>
                              throws <span class="exceptions"><a href="../exceptions/MusialIOException.html" title="class in exceptions">MusialIOException</a></span></div>
<div class="block">Tries to generate the directory specified by the passed <code>File</code> object.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>file</code> - <code>File</code> object representing a directory.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialIOException.html" title="class in exceptions">MusialIOException</a></code> - If the directory could not be generated.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="generateFile(java.io.File)">
<h3>generateFile</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">generateFile</span>&#8203;(<span class="parameters">java.io.File&nbsp;file)</span>
                         throws <span class="exceptions"><a href="../exceptions/MusialIOException.html" title="class in exceptions">MusialIOException</a></span></div>
<div class="block">Tries to generate the file specified by the passed <code>File</code> object.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>file</code> - <code>File</code> object representing a file.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialIOException.html" title="class in exceptions">MusialIOException</a></code> - If the file could not be generated.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="copyFile(java.io.File,java.io.File)">
<h3>copyFile</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">copyFile</span>&#8203;(<span class="parameters">java.io.File&nbsp;file,
java.io.File&nbsp;target)</span>
                     throws <span class="exceptions"><a href="../exceptions/MusialIOException.html" title="class in exceptions">MusialIOException</a></span></div>
<div class="block">Copies the file pointed to with file to the file pointed to with target.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>file</code> - <code>File</code> object, the source file.</dd>
<dd><code>target</code> - <code>File</code> object, the target file.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialIOException.html" title="class in exceptions">MusialIOException</a></code> - If the copy procedure fails.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getSequencesFromPdbStructure(org.biojava.nbio.structure.Structure)">
<h3>getSequencesFromPdbStructure</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">getSequencesFromPdbStructure</span>&#8203;(<span class="parameters">org.biojava.nbio.structure.Structure&nbsp;structure)</span></div>
<div class="block">Returns the chains sequences from a <code>Structure</code> instance parsed from a .pdb file.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>structure</code> - <code>Structure</code> instance to extract the chains sequences from.</dd>
<dt>Returns:</dt>
<dd><code>HashMap</code> storing the .pdb files amino-acid sequences per chain.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="readStructure(java.io.File)">
<h3>readStructure</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">org.biojava.nbio.structure.Structure</span>&nbsp;<span class="member-name">readStructure</span>&#8203;(<span class="parameters">java.io.File&nbsp;pdbFile)</span>
                                                          throws <span class="exceptions">java.io.IOException</span></div>
<div class="block">Parses a <code>Structure</code> instance from a <code>File</code> pointing to a pdb file.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>pdbFile</code> - <code>File</code> instance pointing to a pdb file.</dd>
<dt>Returns:</dt>
<dd><code>Structure</code> object parsed from specified pdb file.</dd>
<dt>Throws:</dt>
<dd><code>java.io.IOException</code> - If any error occurs while parsing the pdb file.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="writeFasta(java.io.File,java.util.HashMap)">
<h3>writeFasta</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">writeFasta</span>&#8203;(<span class="parameters">java.io.File&nbsp;outputFile,
java.util.HashMap&lt;java.lang.String,&#8203;java.util.ArrayList&lt;java.lang.String&gt;&gt;&nbsp;sequences)</span></div>
<div class="block">Writes a fasta format file to the specified output file from a <code>HashMap</code> instance mapping sequences to
 lists of identifiers (used to construct the fasta entry headers).
 <p>
 Each key of the passed map will be used to build one fasta entry. The header of the respective entry is
 constructed by joining all <code>String</code>s of the value accessible via the (sequence) key with the `|` delimiter.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>outputFile</code> - <code>File</code> object pointing to the output fasta file.</dd>
<dd><code>sequences</code> - <code>HashMap</code> mapping sequences to <code>ArrayList</code> of strings.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="writeVcf(java.io.File,java.util.NavigableSet,java.lang.String)">
<h3>writeVcf</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">writeVcf</span>&#8203;(<span class="parameters">java.io.File&nbsp;outputFile,
java.util.NavigableSet&lt;java.lang.String&gt;&nbsp;variants,
java.lang.String&nbsp;chrom)</span></div>
<div class="block">Writes a dummy vcf format file to the specified output file from a <code>NavigableSet</code> instance yielding
 <code>String</code>s that represent variants.
 <p>
 - Each <code>String</code> in the passed <code>NavigableSet</code> has to be formatted as p@r@a, where p is the position,
 r the reference nucleotide content and a the only alternate form content.
 - All variants will be assigned a quality of 1000.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>outputFile</code> - <code>File</code> object pointing to the output vcf file.</dd>
<dd><code>variants</code> - <code>NavigableSet</code> containing variantInformation information, cf. method description.</dd>
<dd><code>chrom</code> - <code>String</code> representing the value to use for the chromosome field of the vcf format for
                   each variantInformation.</dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
</ul>
</section>
<!-- ========= END OF CLASS DATA ========= -->
</main>
<footer role="contentinfo">
<nav role="navigation">
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottom-nav" id="navbar.bottom">
<div class="skip-nav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.bottom.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
</div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<span class="skip-nav" id="skip.navbar.bottom">
<!--   -->
</span></nav>
</footer>
</div>
</div>
</body>
</html>
