<!DOCTYPE HTML>
<html lang="de">
<head>
<!-- Generated by javadoc (17) on Thu Aug 24 09:55:44 CEST 2023 -->
<title>G-Index</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="dc.created" content="2023-08-24">
<meta name="description" content="index: G">
<meta name="generator" content="javadoc/IndexWriter">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
<script type="text/javascript" src="../script-dir/jquery-3.6.1.min.js"></script>
<script type="text/javascript" src="../script-dir/jquery-ui.min.js"></script>
</head>
<body class="index-page">
<script type="text/javascript">var pathtoroot = "../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar-top">
<div class="skip-nav"><a href="#skip-navbar-top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar-top-firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li class="nav-bar-cell1-rev">Index</li>
<li><a href="../help-doc.html#index">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div class="nav-list-search"><label for="search-input">SEARCH:</label>
<input type="text" id="search-input" value="search" disabled="disabled">
<input type="reset" id="reset-button" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip-navbar-top"></span></nav>
</header>
<div class="flex-content">
<main role="main">
<div class="header">
<h1>Index</h1>
</div>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">K</a>&nbsp;<a href="index-11.html">L</a>&nbsp;<a href="index-12.html">M</a>&nbsp;<a href="index-13.html">N</a>&nbsp;<a href="index-14.html">O</a>&nbsp;<a href="index-15.html">P</a>&nbsp;<a href="index-16.html">R</a>&nbsp;<a href="index-17.html">S</a>&nbsp;<a href="index-18.html">T</a>&nbsp;<a href="index-19.html">U</a>&nbsp;<a href="index-20.html">V</a>&nbsp;<a href="index-21.html">W</a>&nbsp;<br><a href="../allclasses-index.html">All&nbsp;Classes&nbsp;and&nbsp;Interfaces</a><span class="vertical-separator">|</span><a href="../allpackages-index.html">All&nbsp;Packages</a><span class="vertical-separator">|</span><a href="../constant-values.html">Constant&nbsp;Field&nbsp;Values</a><span class="vertical-separator">|</span><a href="../serialized-form.html">Serialized&nbsp;Form</a>
<h2 class="title" id="I:G">G</h2>
<dl class="index">
<dt><a href="../utility/Bio.html#GAP" class="member-name-link">GAP</a> - Static variable in class utility.<a href="../utility/Bio.html" title="class in utility">Bio</a></dt>
<dd>
<div class="block">One-letter character used to indicate an alignment gap.</div>
</dd>
<dt><a href="../utility/IO.html#generateDirectory(java.io.File)" class="member-name-link">generateDirectory(File)</a> - Static method in class utility.<a href="../utility/IO.html" title="class in utility">IO</a></dt>
<dd>
<div class="block">Tries to generate the directory specified by the passed <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/File.html" title="class or interface in java.io" class="external-link"><code>File</code></a> object.</div>
</dd>
<dt><a href="../utility/IO.html#generateFile(java.io.File)" class="member-name-link">generateFile(File)</a> - Static method in class utility.<a href="../utility/IO.html" title="class in utility">IO</a></dt>
<dd>
<div class="block">Tries to generate the file specified by the passed <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/File.html" title="class or interface in java.io" class="external-link"><code>File</code></a> object.</div>
</dd>
<dt><a href="../datastructure/Feature.html#getAllele(java.lang.String)" class="member-name-link">getAllele(String)</a> - Method in class datastructure.<a href="../datastructure/Feature.html" title="class in datastructure">Feature</a></dt>
<dd>
<div class="block">Returns the <a href="../datastructure/Form.html" title="class in datastructure"><code>Form</code></a>, i.e., allele, stored with alleleName from <a href="../datastructure/Feature.html#alleles"><code>Feature.alleles</code></a> or null if it does not exist.</div>
</dd>
<dt><a href="../datastructure/Feature.html#getAlleleCount()" class="member-name-link">getAlleleCount()</a> - Method in class datastructure.<a href="../datastructure/Feature.html" title="class in datastructure">Feature</a></dt>
<dd>
<div class="block">Returns the number of alleles stored for this feature.</div>
</dd>
<dt><a href="../datastructure/Feature.html#getAlleleNameIterator()" class="member-name-link">getAlleleNameIterator()</a> - Method in class datastructure.<a href="../datastructure/Feature.html" title="class in datastructure">Feature</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../datastructure/FeatureCoding.html#getAminoacidVariantAnnotation(int,java.lang.String)" class="member-name-link">getAminoacidVariantAnnotation(int, String)</a> - Method in class datastructure.<a href="../datastructure/FeatureCoding.html" title="class in datastructure">FeatureCoding</a></dt>
<dd>
<div class="block">Returns the <a href="../datastructure/VariantAnnotation.html" title="class in datastructure"><code>VariantAnnotation</code></a> stored at this instances <a href="../datastructure/FeatureCoding.html#aminoacidVariants"><code>FeatureCoding.aminoacidVariants</code></a> at position and content.</div>
</dd>
<dt><a href="../datastructure/FeatureCoding.html#getAminoacidVariantPositions()" class="member-name-link">getAminoacidVariantPositions()</a> - Method in class datastructure.<a href="../datastructure/FeatureCoding.html" title="class in datastructure">FeatureCoding</a></dt>
<dd>
<div class="block">Retrieves a set of all variant positions of this <a href="../datastructure/FeatureCoding.html" title="class in datastructure"><code>FeatureCoding</code></a>.</div>
</dd>
<dt><a href="../datastructure/FeatureCoding.html#getAminoacidVariants(int)" class="member-name-link">getAminoacidVariants(int)</a> - Method in class datastructure.<a href="../datastructure/FeatureCoding.html" title="class in datastructure">FeatureCoding</a></dt>
<dd>
<div class="block">Returns all <a href="../datastructure/VariantAnnotation.html" title="class in datastructure"><code>VariantAnnotation</code></a>s stored at this instances <a href="../datastructure/FeatureCoding.html#aminoacidVariants"><code>FeatureCoding.aminoacidVariants</code></a> at position.</div>
</dd>
<dt><a href="../datastructure/FeatureCoding.html#getAminoacidVariants(java.lang.String)" class="member-name-link">getAminoacidVariants(String)</a> - Method in class datastructure.<a href="../datastructure/FeatureCoding.html" title="class in datastructure">FeatureCoding</a></dt>
<dd>
<div class="block">Returns a map representation of all variants of a single proteoform of this <a href="../datastructure/FeatureCoding.html" title="class in datastructure"><code>FeatureCoding</code></a>.</div>
</dd>
<dt><a href="../datastructure/Feature.html#getAnnotation(java.lang.String)" class="member-name-link">getAnnotation(String)</a> - Method in class datastructure.<a href="../datastructure/Feature.html" title="class in datastructure">Feature</a></dt>
<dd>
<div class="block">Returns the value stored under key in <a href="../datastructure/Feature.html#annotations"><code>Feature.annotations</code></a>.</div>
</dd>
<dt><a href="../datastructure/Form.html#getAnnotation(java.lang.String)" class="member-name-link">getAnnotation(String)</a> - Method in class datastructure.<a href="../datastructure/Form.html" title="class in datastructure">Form</a></dt>
<dd>
<div class="block">Query an annotation value from <a href="../datastructure/Form.html#annotations"><code>Form.annotations</code></a> or null, if it does not exist.</div>
</dd>
<dt><a href="../datastructure/Sample.html#getAnnotation(java.lang.String)" class="member-name-link">getAnnotation(String)</a> - Method in class datastructure.<a href="../datastructure/Sample.html" title="class in datastructure">Sample</a></dt>
<dd>
<div class="block">Query an annotation value from <a href="../datastructure/Sample.html#annotations"><code>Sample.annotations</code></a> or null, if it does not exist.</div>
</dd>
<dt><a href="../datastructure/Feature.html#getAnnotations()" class="member-name-link">getAnnotations()</a> - Method in class datastructure.<a href="../datastructure/Feature.html" title="class in datastructure">Feature</a></dt>
<dd>
<div class="block">Returns all annotations of this instance.</div>
</dd>
<dt><a href="../datastructure/Sample.html#getAnnotations()" class="member-name-link">getAnnotations()</a> - Method in class datastructure.<a href="../datastructure/Sample.html" title="class in datastructure">Sample</a></dt>
<dd>
<div class="block">Queries all annotation values from <a href="../datastructure/Sample.html#annotations"><code>Sample.annotations</code></a>.</div>
</dd>
<dt><a href="../datastructure/FeatureCoding.html#getCodingSequence()" class="member-name-link">getCodingSequence()</a> - Method in class datastructure.<a href="../datastructure/FeatureCoding.html" title="class in datastructure">FeatureCoding</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../datastructure/MusialStorage.html#getFeature(java.lang.String)" class="member-name-link">getFeature(String)</a> - Method in class datastructure.<a href="../datastructure/MusialStorage.html" title="class in datastructure">MusialStorage</a></dt>
<dd>
<div class="block">Getter method for <a href="../datastructure/MusialStorage.html#features"><code>MusialStorage.features</code></a>.</div>
</dd>
<dt><a href="../datastructure/MusialStorage.html#getFeatureNameIterator()" class="member-name-link">getFeatureNameIterator()</a> - Method in class datastructure.<a href="../datastructure/MusialStorage.html" title="class in datastructure">MusialStorage</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../datastructure/MusialStorage.html#getFeatureNameIterator(int)" class="member-name-link">getFeatureNameIterator(int)</a> - Method in class datastructure.<a href="../datastructure/MusialStorage.html" title="class in datastructure">MusialStorage</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../datastructure/Feature.html#getNucleotideVariantAnnotation(int,java.lang.String)" class="member-name-link">getNucleotideVariantAnnotation(int, String)</a> - Method in class datastructure.<a href="../datastructure/Feature.html" title="class in datastructure">Feature</a></dt>
<dd>
<div class="block">Returns the <a href="../datastructure/VariantAnnotation.html" title="class in datastructure"><code>VariantAnnotation</code></a> stored at this instances <a href="../datastructure/Feature.html#nucleotideVariants"><code>Feature.nucleotideVariants</code></a> at position and content.</div>
</dd>
<dt><a href="../datastructure/Feature.html#getNucleotideVariantPositions()" class="member-name-link">getNucleotideVariantPositions()</a> - Method in class datastructure.<a href="../datastructure/Feature.html" title="class in datastructure">Feature</a></dt>
<dd>
<div class="block">Retrieves a set of all variant positions of this <a href="../datastructure/Feature.html" title="class in datastructure"><code>Feature</code></a>.</div>
</dd>
<dt><a href="../datastructure/Feature.html#getNucleotideVariants(int)" class="member-name-link">getNucleotideVariants(int)</a> - Method in class datastructure.<a href="../datastructure/Feature.html" title="class in datastructure">Feature</a></dt>
<dd>
<div class="block">Returns all <a href="../datastructure/VariantAnnotation.html" title="class in datastructure"><code>VariantAnnotation</code></a>s stored at this instances <a href="../datastructure/Feature.html#nucleotideVariants"><code>Feature.nucleotideVariants</code></a> at position.</div>
</dd>
<dt><a href="../datastructure/Feature.html#getNucleotideVariants(java.lang.String)" class="member-name-link">getNucleotideVariants(String)</a> - Method in class datastructure.<a href="../datastructure/Feature.html" title="class in datastructure">Feature</a></dt>
<dd>
<div class="block">Returns a map representation of all variants of a single allele of this <a href="../datastructure/Feature.html" title="class in datastructure"><code>Feature</code></a>.</div>
</dd>
<dt><a href="../datastructure/MusialStorage.html#getNumberOfSamples()" class="member-name-link">getNumberOfSamples()</a> - Method in class datastructure.<a href="../datastructure/MusialStorage.html" title="class in datastructure">MusialStorage</a></dt>
<dd>
<div class="block">Getter method for <a href="../datastructure/MusialStorage.html#samples"><code>MusialStorage.samples</code></a> size.</div>
</dd>
<dt><a href="../datastructure/Form.html#getOccurrence()" class="member-name-link">getOccurrence()</a> - Method in class datastructure.<a href="../datastructure/Form.html" title="class in datastructure">Form</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../datastructure/VariantAnnotation.html#getProperty(java.lang.String)" class="member-name-link">getProperty(String)</a> - Method in class datastructure.<a href="../datastructure/VariantAnnotation.html" title="class in datastructure">VariantAnnotation</a></dt>
<dd>
<div class="block">Retrieves the value stored under key in this <a href="../datastructure/VariantAnnotation.html#properties"><code>properties</code></a>.</div>
</dd>
<dt><a href="../datastructure/VariantAnnotation.html#getPropertyKeys()" class="member-name-link">getPropertyKeys()</a> - Method in class datastructure.<a href="../datastructure/VariantAnnotation.html" title="class in datastructure">VariantAnnotation</a></dt>
<dd>
<div class="block">Retrieves the keys stored in this <a href="../datastructure/VariantAnnotation.html#properties"><code>properties</code></a>.</div>
</dd>
<dt><a href="../datastructure/FeatureCoding.html#getProteoform(java.lang.String)" class="member-name-link">getProteoform(String)</a> - Method in class datastructure.<a href="../datastructure/FeatureCoding.html" title="class in datastructure">FeatureCoding</a></dt>
<dd>
<div class="block">Returns the <a href="../datastructure/Form.html" title="class in datastructure"><code>Form</code></a>, i.e., proteoform, stored with proteoformName from <a href="../datastructure/FeatureCoding.html#proteoforms"><code>FeatureCoding.proteoforms</code></a> or null if it does not exist.</div>
</dd>
<dt><a href="../datastructure/FeatureCoding.html#getProteoformCount()" class="member-name-link">getProteoformCount()</a> - Method in class datastructure.<a href="../datastructure/FeatureCoding.html" title="class in datastructure">FeatureCoding</a></dt>
<dd>
<div class="block">Returns the number of proteoforms stored for this feature.</div>
</dd>
<dt><a href="../datastructure/FeatureCoding.html#getProteoformNameIterator()" class="member-name-link">getProteoformNameIterator()</a> - Method in class datastructure.<a href="../datastructure/FeatureCoding.html" title="class in datastructure">FeatureCoding</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../datastructure/MusialStorage.html#getReferenceSequenceOfFeature(java.lang.String)" class="member-name-link">getReferenceSequenceOfFeature(String)</a> - Method in class datastructure.<a href="../datastructure/MusialStorage.html" title="class in datastructure">MusialStorage</a></dt>
<dd>
<div class="block">Constructs the base sequence of the specified feature wrt.</div>
</dd>
<dt><a href="../datastructure/MusialStorage.html#getSample(java.lang.String)" class="member-name-link">getSample(String)</a> - Method in class datastructure.<a href="../datastructure/MusialStorage.html" title="class in datastructure">MusialStorage</a></dt>
<dd>
<div class="block">Getter method for <a href="../datastructure/MusialStorage.html#samples"><code>MusialStorage.samples</code></a>.</div>
</dd>
<dt><a href="../datastructure/MusialStorage.html#getSampleNameIterator()" class="member-name-link">getSampleNameIterator()</a> - Method in class datastructure.<a href="../datastructure/MusialStorage.html" title="class in datastructure">MusialStorage</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../utility/IO.html#getSequencesFromPdbStructure(org.biojava.nbio.structure.Structure)" class="member-name-link">getSequencesFromPdbStructure(Structure)</a> - Static method in class utility.<a href="../utility/IO.html" title="class in utility">IO</a></dt>
<dd>
<div class="block">Returns the chains sequences from a <code>Structure</code> instance parsed from a .pdb file.</div>
</dd>
<dt><a href="../datastructure/FeatureCoding.html#getStructure()" class="member-name-link">getStructure()</a> - Method in class datastructure.<a href="../datastructure/FeatureCoding.html" title="class in datastructure">FeatureCoding</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../utility/Logger.html#getTimestampDayTime()" class="member-name-link">getTimestampDayTime()</a> - Static method in class utility.<a href="../utility/Logger.html" title="class in utility">Logger</a></dt>
<dd>
<div class="block">Generates, formats and returns a <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link"><code>String</code></a> representing the current date and time.</div>
</dd>
<dt><a href="../runnables/SampleAnalyzer.html#getVariantCallsIterator()" class="member-name-link">getVariantCallsIterator()</a> - Method in class runnables.<a href="../runnables/SampleAnalyzer.html" title="class in runnables">SampleAnalyzer</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../utility/Bio.html#getVariantsOfAlignedSequences(java.lang.String,java.lang.String)" class="member-name-link">getVariantsOfAlignedSequences(String, String)</a> - Static method in class utility.<a href="../utility/Bio.html" title="class in utility">Bio</a></dt>
<dd>
<div class="block">Computes variants from two aligned sequences.</div>
</dd>
<dt><a href="../utility/Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html#%3Cinit%3E()" class="member-name-link">GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES()</a> - Constructor for enum class utility.<a href="../utility/Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum class in utility">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../utility/Bio.html#globalAminoAcidSequenceAlignment(java.lang.String,java.lang.String,int,int,utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,java.lang.Integer)" class="member-name-link">globalAminoAcidSequenceAlignment(String, String, int, int, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, Integer)</a> - Static method in class utility.<a href="../utility/Bio.html" title="class in utility">Bio</a></dt>
<dd>
<div class="block">Invokes the <a href="../utility/Bio.html#globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int%5B%5D%5B%5D,int,int,utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,java.lang.Integer)"><code>Bio.globalSequenceAlignment(java.lang.String, java.lang.String, java.util.HashMap&lt;java.lang.Character, java.lang.Integer&gt;, int[][], int, int, utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, java.lang.Integer)</code></a> method with pre-specified parameters for amino-acid sequence
 alignment.</div>
</dd>
<dt><a href="../utility/Bio.html#globalNucleotideSequenceAlignment(java.lang.String,java.lang.String,utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,java.lang.Integer)" class="member-name-link">globalNucleotideSequenceAlignment(String, String, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, Integer)</a> - Static method in class utility.<a href="../utility/Bio.html" title="class in utility">Bio</a></dt>
<dd>
<div class="block">Invokes the <a href="../utility/Bio.html#globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int%5B%5D%5B%5D,int,int,utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,java.lang.Integer)"><code>Bio.globalSequenceAlignment(java.lang.String, java.lang.String, java.util.HashMap&lt;java.lang.Character, java.lang.Integer&gt;, int[][], int, int, utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, java.lang.Integer)</code></a> method with pre-specified parameters for nucleotide sequence
 alignment.</div>
</dd>
<dt><a href="../utility/Bio.html#globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int%5B%5D%5B%5D,int,int,utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,utility.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,java.lang.Integer)" class="member-name-link">globalSequenceAlignment(String, String, HashMap&lt;Character, Integer&gt;, int[][], int, int, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, Integer)</a> - Static method in class utility.<a href="../utility/Bio.html" title="class in utility">Bio</a></dt>
<dd>
<div class="block">Computes a global sequence alignment using the gap-affine Needleman-Wunsch algorithm.</div>
</dd>
</dl>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">K</a>&nbsp;<a href="index-11.html">L</a>&nbsp;<a href="index-12.html">M</a>&nbsp;<a href="index-13.html">N</a>&nbsp;<a href="index-14.html">O</a>&nbsp;<a href="index-15.html">P</a>&nbsp;<a href="index-16.html">R</a>&nbsp;<a href="index-17.html">S</a>&nbsp;<a href="index-18.html">T</a>&nbsp;<a href="index-19.html">U</a>&nbsp;<a href="index-20.html">V</a>&nbsp;<a href="index-21.html">W</a>&nbsp;<br><a href="../allclasses-index.html">All&nbsp;Classes&nbsp;and&nbsp;Interfaces</a><span class="vertical-separator">|</span><a href="../allpackages-index.html">All&nbsp;Packages</a><span class="vertical-separator">|</span><a href="../constant-values.html">Constant&nbsp;Field&nbsp;Values</a><span class="vertical-separator">|</span><a href="../serialized-form.html">Serialized&nbsp;Form</a></main>
</div>
</div>
</body>
</html>
