<!DOCTYPE HTML>
<!-- NewPage -->
<html lang="de">
<head>
<!-- Generated by javadoc (15) on Mon Jul 25 13:17:44 CEST 2022 -->
<title>G-Index</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="dc.created" content="2022-07-25">
<meta name="description" content="index: G">
<meta name="generator" content="javadoc/SplitIndexWriter">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
<script type="text/javascript" src="../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../script-dir/jquery-ui.min.js"></script>
</head>
<body class="split-index-page">
<script type="text/javascript">var pathtoroot = "../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar.top">
<div class="skip-nav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.top.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li class="nav-bar-cell1-rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div class="nav-list-search"><label for="search">SEARCH:</label>
<input type="text" id="search" value="search" disabled="disabled">
<input type="reset" id="reset" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip.navbar.top">
<!--   -->
</span></nav>
</header>
<div class="flex-content">
<main role="main">
<div class="header">
<h1>Index</h1>
</div>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">L</a>&nbsp;<a href="index-11.html">M</a>&nbsp;<a href="index-12.html">N</a>&nbsp;<a href="index-13.html">O</a>&nbsp;<a href="index-14.html">P</a>&nbsp;<a href="index-15.html">R</a>&nbsp;<a href="index-16.html">S</a>&nbsp;<a href="index-17.html">T</a>&nbsp;<a href="index-18.html">U</a>&nbsp;<a href="index-19.html">V</a>&nbsp;<a href="index-20.html">W</a>&nbsp;<br><a href="../allclasses-index.html">All&nbsp;Classes</a><span class="vertical-separator">|</span><a href="../allpackages-index.html">All&nbsp;Packages</a>
<h2 class="title" id="I:G">G</h2>
<dl class="index">
<dt><span class="member-name-link"><a href="../components/Bio.html#GAP">GAP</a></span> - Static variable in class components.<a href="../components/Bio.html" title="class in components">Bio</a></dt>
<dd>
<div class="block">One-letter character used to indicate an alignment gap.</div>
</dd>
<dt><span class="member-name-link"><a href="../components/IO.html#generateDirectory(java.io.File)">generateDirectory(File)</a></span> - Static method in class components.<a href="../components/IO.html" title="class in components">IO</a></dt>
<dd>
<div class="block">Tries to generate the directory specified by the passed <code>File</code> object.</div>
</dd>
<dt><span class="member-name-link"><a href="../components/IO.html#generateFile(java.io.File)">generateFile(File)</a></span> - Static method in class components.<a href="../components/IO.html" title="class in components">IO</a></dt>
<dd>
<div class="block">Tries to generate the file specified by the passed <code>File</code> object.</div>
</dd>
<dt><span class="member-name-link"><a href="../datastructure/AllocatedProteinEntry.html#generateProteoformName(java.lang.String)">generateProteoformName(String)</a></span> - Method in class datastructure.<a href="../datastructure/AllocatedProteinEntry.html" title="class in datastructure">AllocatedProteinEntry</a></dt>
<dd>
<div class="block">Returns a <code>String</code> that is used as internal proteoform name.</div>
</dd>
<dt><span class="member-name-link"><a href="../components/Logging.html#getDoneMessage()">getDoneMessage()</a></span> - Static method in class components.<a href="../components/Logging.html" title="class in components">Logging</a></dt>
<dd>
<div class="block">Generates a <code>String</code> stating "Done" concatenated with time-stamp information.</div>
</dd>
<dt><span class="member-name-link"><a href="../datastructure/FastaContainer.html#getHeader()">getHeader()</a></span> - Method in class datastructure.<a href="../datastructure/FastaContainer.html" title="class in datastructure">FastaContainer</a></dt>
<dd>
<div class="block">Returns the stored header line.</div>
</dd>
<dt><span class="member-name-link"><a href="../datastructure/VariantsDictionary.html#getNucleotideSequence(java.lang.String,java.lang.String)">getNucleotideSequence(String, String)</a></span> - Method in class datastructure.<a href="../datastructure/VariantsDictionary.html" title="class in datastructure">VariantsDictionary</a></dt>
<dd>
<div class="block">Extracts a <code>String</code> yielding the nucleotide sequence with all variants of one sample being incorporated for one feature.</div>
</dd>
<dt><span class="member-name-link"><a href="../datastructure/VariantsDictionary.html#getProteoformSequence(java.lang.String,java.lang.String)">getProteoformSequence(String, String)</a></span> - Method in class datastructure.<a href="../datastructure/VariantsDictionary.html" title="class in datastructure">VariantsDictionary</a></dt>
<dd>
<div class="block">Extracts a <code>String</code> yielding the amino-acid sequence with all variants of one proteoform being incorporated
 for one feature.</div>
</dd>
<dt><span class="member-name-link"><a href="../datastructure/VariantsDictionary.html#getProteoformVariants(java.lang.String,java.lang.String,boolean)">getProteoformVariants(String, String, boolean)</a></span> - Method in class datastructure.<a href="../datastructure/VariantsDictionary.html" title="class in datastructure">VariantsDictionary</a></dt>
<dd>
<div class="block">Extracts a <code>HashMap</code> of <code>String</code>/<code>String</code> key/value pairs reflecting all (filtered) variants
 for one proteoform and feature.</div>
</dd>
<dt><span class="member-name-link"><a href="../datastructure/VariantsDictionary.html#getSampleVariants(java.lang.String,java.lang.String)">getSampleVariants(String, String)</a></span> - Method in class datastructure.<a href="../datastructure/VariantsDictionary.html" title="class in datastructure">VariantsDictionary</a></dt>
<dd>
<div class="block">Extracts a <code>HashMap</code> of <code>Integer</code>/<code>String</code> key/value pairs reflecting all (filtered) variants
 for one sample and feature.</div>
</dd>
<dt><span class="member-name-link"><a href="../datastructure/FastaContainer.html#getSequence()">getSequence()</a></span> - Method in class datastructure.<a href="../datastructure/FastaContainer.html" title="class in datastructure">FastaContainer</a></dt>
<dd>
<div class="block">Returns the stored sequence in full length.</div>
</dd>
<dt><span class="member-name-link"><a href="../datastructure/FastaContainer.html#getSequence(int,int)">getSequence(int, int)</a></span> - Method in class datastructure.<a href="../datastructure/FastaContainer.html" title="class in datastructure">FastaContainer</a></dt>
<dd>
<div class="block">Returns a sub sequence of the stored sequence.</div>
</dd>
<dt><span class="member-name-link"><a href="../components/IO.html#getSequencesFromPdbStructure(org.biojava.nbio.structure.Structure)">getSequencesFromPdbStructure(Structure)</a></span> - Static method in class components.<a href="../components/IO.html" title="class in components">IO</a></dt>
<dd>
<div class="block">Returns the chains sequences from a <code>Structure</code> instance parsed from a .pdb file.</div>
</dd>
<dt><span class="member-name-link"><a href="../components/Logging.html#getTimestampDayTime()">getTimestampDayTime()</a></span> - Static method in class components.<a href="../components/Logging.html" title="class in components">Logging</a></dt>
<dd>
<div class="block">Generates, formats and returns a <code>String</code> representing the current date and time.</div>
</dd>
<dt><span class="member-name-link"><a href="../components/Logging.html#getTimestampTime()">getTimestampTime()</a></span> - Static method in class components.<a href="../components/Logging.html" title="class in components">Logging</a></dt>
<dd>
<div class="block">Generates, formats and returns a <code>String</code> representing the current time.</div>
</dd>
<dt><span class="member-name-link"><a href="../components/Bio.html#getVariantsOfAlignedSequences(java.lang.String,java.lang.String)">getVariantsOfAlignedSequences(String, String)</a></span> - Static method in class components.<a href="../components/Bio.html" title="class in components">Bio</a></dt>
<dd>
<div class="block">Computes variants from two aligned sequences.</div>
</dd>
<dt><span class="member-name-link"><a href="../components/Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html#%3Cinit%3E()">GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES()</a></span> - Constructor for enum components.<a href="../components/Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES.html" title="enum in components">Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES</a></dt>
<dd>&nbsp;</dd>
<dt><span class="member-name-link"><a href="../components/Bio.html#globalAminoAcidSequenceAlignment(java.lang.String,java.lang.String,int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)">globalAminoAcidSequenceAlignment(String, String, int, int, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)</a></span> - Static method in class components.<a href="../components/Bio.html" title="class in components">Bio</a></dt>
<dd>
<div class="block">Invokes the <a href="../components/Bio.html#globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int%5B%5D%5B%5D,int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)"><code>Bio.globalSequenceAlignment(java.lang.String, java.lang.String, java.util.HashMap&lt;java.lang.Character, java.lang.Integer&gt;, int[][], int, int, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)</code></a> method with pre-specified parameters for amino-acid sequence
 alignment.</div>
</dd>
<dt><span class="member-name-link"><a href="../components/Bio.html#globalNucleotideSequenceAlignment(java.lang.String,java.lang.String,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)">globalNucleotideSequenceAlignment(String, String, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)</a></span> - Static method in class components.<a href="../components/Bio.html" title="class in components">Bio</a></dt>
<dd>
<div class="block">Invokes the <a href="../components/Bio.html#globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int%5B%5D%5B%5D,int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)"><code>Bio.globalSequenceAlignment(java.lang.String, java.lang.String, java.util.HashMap&lt;java.lang.Character, java.lang.Integer&gt;, int[][], int, int, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)</code></a> method with pre-specified parameters for nucleotide sequence
 alignment.</div>
</dd>
<dt><span class="member-name-link"><a href="../components/Bio.html#globalSequenceAlignment(java.lang.String,java.lang.String,java.util.HashMap,int%5B%5D%5B%5D,int,int,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES,components.Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)">globalSequenceAlignment(String, String, HashMap&lt;Character, Integer&gt;, int[][], int, int, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES, Bio.GLOBAL_SEQUENCE_ALIGNMENT_MARGIN_GAP_MODES)</a></span> - Static method in class components.<a href="../components/Bio.html" title="class in components">Bio</a></dt>
<dd>
<div class="block">Computes a global sequence alignment using the gap-affine Needleman-Wunsch algorithm.</div>
</dd>
<dt><span class="member-name-link"><a href="../cli/CLIParametersInferSequences.html#GS">GS</a></span> - Variable in class cli.<a href="../cli/CLIParametersInferSequences.html" title="class in cli">CLIParametersInferSequences</a></dt>
<dd>
<div class="block">Whether to output genotype sequences.</div>
</dd>
<dt><span class="member-name-link"><a href="../cli/CLIParametersInferSequences.html#GSMSA">GSMSA</a></span> - Variable in class cli.<a href="../cli/CLIParametersInferSequences.html" title="class in cli">CLIParametersInferSequences</a></dt>
<dd>
<div class="block">Whether to output genotype sequences as multiple sequence alignment.</div>
</dd>
</dl>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">L</a>&nbsp;<a href="index-11.html">M</a>&nbsp;<a href="index-12.html">N</a>&nbsp;<a href="index-13.html">O</a>&nbsp;<a href="index-14.html">P</a>&nbsp;<a href="index-15.html">R</a>&nbsp;<a href="index-16.html">S</a>&nbsp;<a href="index-17.html">T</a>&nbsp;<a href="index-18.html">U</a>&nbsp;<a href="index-19.html">V</a>&nbsp;<a href="index-20.html">W</a>&nbsp;<br><a href="../allclasses-index.html">All&nbsp;Classes</a><span class="vertical-separator">|</span><a href="../allpackages-index.html">All&nbsp;Packages</a></main>
<footer role="contentinfo">
<nav role="navigation">
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottom-nav" id="navbar.bottom">
<div class="skip-nav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.bottom.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li class="nav-bar-cell1-rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<span class="skip-nav" id="skip.navbar.bottom">
<!--   -->
</span></nav>
</footer>
</div>
</div>
</body>
</html>
