<!DOCTYPE HTML>
<!-- NewPage -->
<html lang="de">
<head>
<!-- Generated by javadoc (15) on Tue Dec 07 18:26:26 CET 2021 -->
<title>Bio</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="dc.created" content="2021-12-07">
<meta name="description" content="declaration: package: utility, class: Bio">
<meta name="generator" content="javadoc/ClassWriterImpl">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
<script type="text/javascript" src="../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../script-dir/jquery-ui.min.js"></script>
</head>
<body class="class-declaration-page">
<script type="text/javascript">var data = {"i0":9,"i1":9,"i2":9,"i3":9,"i4":9,"i5":9,"i6":9,"i7":9,"i8":9,"i9":9,"i10":9,"i11":9,"i12":9,"i13":9};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],8:["t4","Concrete Methods"]};
var altColor = "alt-color";
var rowColor = "row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar.top">
<div class="skip-nav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.top.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search">SEARCH:</label>
<input type="text" id="search" value="search" disabled="disabled">
<input type="reset" id="reset" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip.navbar.top">
<!--   -->
</span></nav>
</header>
<div class="flex-content">
<main role="main">
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="sub-title"><span class="package-label-in-type">Package</span>&nbsp;<a href="package-summary.html">utility</a></div>
<h1 title="Class Bio" class="title">Class Bio</h1>
</div>
<div class="inheritance" title="Inheritance Tree">java.lang.Object
<div class="inheritance">utility.Bio</div>
</div>
<section class="description">
<hr>
<pre>public final class <span class="type-name-label">Bio</span>
extends java.lang.Object</pre>
<div class="block">Comprises static methods used to solve specifically bioinformatics problems.</div>
<dl class="notes">
<dt>Since:</dt>
<dd>2.0</dd>
<dt>Version:</dt>
<dd>2.0</dd>
<dt>Author:</dt>
<dd>Simon Hackl</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<!-- =========== FIELD SUMMARY =========== -->
<li>
<section class="field-summary" id="field.summary">
<h2>Field Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Fields</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Field</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.Character,&#8203;java.lang.Integer&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#AA1_BLOSUM80_INDEX">AA1_BLOSUM80_INDEX</a></span></code></th>
<td class="col-last">
<div class="block">Hash map mapping amino-acid one letter symbols to indices used to access substitution matrix values.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#AA1TO3">AA1TO3</a></span></code></th>
<td class="col-last">
<div class="block">Hash map mapping amino-acid one to three letter codes.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#AMBIGUOUS_AA3">AMBIGUOUS_AA3</a></span></code></th>
<td class="col-last">
<div class="block">Three letter code string used to indicate an ambiguous translation of a codon (i.e. due to an translational
 shift induced by indels).</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>private static int[][]</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#BLOSUM80">BLOSUM80</a></span></code></th>
<td class="col-last">
<div class="block">The BLOSUM80 matrix used for alignment computation.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#CODON_MAP">CODON_MAP</a></span></code></th>
<td class="col-last">
<div class="block">Hash map mapping nucleotide codons to single-letter amino acids.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>private static int</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#GAP_EXTENSION_PENALTY">GAP_EXTENSION_PENALTY</a></span></code></th>
<td class="col-last">
<div class="block">Gap extension penalty used for alignment computation.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>private static int</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#GAP_OPEN_PENALTY">GAP_OPEN_PENALTY</a></span></code></th>
<td class="col-last">
<div class="block">Gap opening penalty used for alignment computation.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#NO_MATCH_CHAR">NO_MATCH_CHAR</a></span></code></th>
<td class="col-last">
<div class="block">Character to indicate a gap/missing information regarding a nucleotide and amino acid sequence.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#SEPARATOR_CHAR">SEPARATOR_CHAR</a></span></code></th>
<td class="col-last">
<div class="block">Character used as a separator in padded sequences.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#TERMINATION_AA1">TERMINATION_AA1</a></span></code></th>
<td class="col-last">
<div class="block">One letter code character used to indicate a translated stop codon.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#TERMINATION_AA3">TERMINATION_AA3</a></span></code></th>
<td class="col-last">
<div class="block">Three letter code String used to indicate a translated stop codon.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#TRUNCATED_CHAR">TRUNCATED_CHAR</a></span></code></th>
<td class="col-last">
<div class="block">Character to indicate a truncation of a sequence.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#UNKNOWN_AA1">UNKNOWN_AA1</a></span></code></th>
<td class="col-last">
<div class="block">One letter code character used to indicate the translation of a codon containing an N character (i.e. unknown or
 any nucleotide).</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#UNKNOWN_AA3">UNKNOWN_AA3</a></span></code></th>
<td class="col-last">
<div class="block">Three letter code string used to indicate the translation of a codon containing an N character (i.e. unknown or
 any nucleotide).</div>
</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<li>
<section class="constructor-summary" id="constructor.summary">
<h2>Constructor Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Constructors</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Constructor</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<th class="col-constructor-name" scope="row"><code><span class="member-name-link"><a href="#%3Cinit%3E()">Bio</a></span>()</code></th>
<td class="col-last">&nbsp;</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ========== METHOD SUMMARY =========== -->
<li>
<section class="method-summary" id="method.summary">
<h2>Method Summary</h2>
<div class="member-summary" id="method-summary-table">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button role="tab" aria-selected="true" aria-controls="method-summary-table.tabpanel" tabindex="0" onkeydown="switchTab(event)" id="t0" class="active-table-tab">All Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t1" class="table-tab" onclick="show(1);">Static Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t4" class="table-tab" onclick="show(8);">Concrete Methods</button></div>
<div id="method-summary-table.tabpanel" role="tabpanel">
<table class="summary-table" aria-labelledby="t0">
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Method</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color" id="i0">
<td class="col-first"><code>static int</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getFirstResidueNumberFromProteinStructure(org.biojava.nbio.structure.Structure,java.lang.String)">getFirstResidueNumberFromProteinStructure</a></span>&#8203;(org.biojava.nbio.structure.Structure&nbsp;structure,
java.lang.String&nbsp;chainAsymId)</code></th>
<td class="col-last">
<div class="block">Returns the first residue number from a <code>Structure</code> instances specified chain (via chainId).</div>
</td>
</tr>
<tr class="row-color" id="i1">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.Integer,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getInsertionsOfSample(datastructure.VariantPositionsTable,java.lang.String,java.lang.String,java.util.NavigableSet,boolean)">getInsertionsOfSample</a></span>&#8203;(<a href="../datastructure/VariantPositionsTable.html" title="class in datastructure">VariantPositionsTable</a>&nbsp;table,
java.lang.String&nbsp;referenceAnalysisId,
java.lang.String&nbsp;sampleName,
java.util.NavigableSet&lt;java.lang.String&gt;&nbsp;positions,
boolean&nbsp;asSense)</code></th>
<td class="col-last">
<div class="block">Constructs and returns a <code>HashMap</code> mapping <code>String</code> representations of genome positions to
 <code>String</code>s representing inserted nucleotide sequences with regard to one analyzed sample.</div>
</td>
</tr>
<tr class="alt-color" id="i2">
<td class="col-first"><code>static org.javatuples.Pair&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getNucleotideToAminoAcidSequenceMatching(java.lang.String,java.lang.String)">getNucleotideToAminoAcidSequenceMatching</a></span>&#8203;(java.lang.String&nbsp;aminoAcidSequence,
java.lang.String&nbsp;nucleotideSequence)</code></th>
<td class="col-last">
<div class="block">Matches a amino-acid and nucleotide sequence.</div>
</td>
</tr>
<tr class="row-color" id="i3">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;org.biojava.nbio.structure.Group&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getResiduesFromStructure(org.biojava.nbio.structure.Structure)">getResiduesFromStructure</a></span>&#8203;(org.biojava.nbio.structure.Structure&nbsp;structure)</code></th>
<td class="col-last">
<div class="block">Returns a <code>HashMap</code> mapping residue number and chain names as (concatenated) keys to the respective
 <code>Group</code> instances.</div>
</td>
</tr>
<tr class="alt-color" id="i4">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getReverseComplement(java.lang.String)">getReverseComplement</a></span>&#8203;(java.lang.String&nbsp;sequence)</code></th>
<td class="col-last">
<div class="block">Returns the reverse complement of the passed nucleotide sequence.</div>
</td>
</tr>
<tr class="row-color" id="i5">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getSequenceFromTable(datastructure.VariantPositionsTable,java.lang.String,java.lang.String,java.util.Iterator,boolean,boolean)">getSequenceFromTable</a></span>&#8203;(<a href="../datastructure/VariantPositionsTable.html" title="class in datastructure">VariantPositionsTable</a>&nbsp;table,
java.lang.String&nbsp;referenceAnalysisId,
java.lang.String&nbsp;sampleName,
java.util.Iterator&lt;java.lang.String&gt;&nbsp;positions,
boolean&nbsp;addDeletions,
boolean&nbsp;addInsertions)</code></th>
<td class="col-last">
<div class="block">Constructs and returns a nucleotide sequence as <code>String</code> for the specified sample by using the specified
 variants table,
 reference feature and positions.</div>
</td>
</tr>
<tr class="alt-color" id="i6">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getSequencesFromPdb(org.biojava.nbio.structure.Structure)">getSequencesFromPdb</a></span>&#8203;(org.biojava.nbio.structure.Structure&nbsp;structure)</code></th>
<td class="col-last">
<div class="block">Returns the chains sequences from a <code>Structure</code> instance parsed from a .pdb file.</div>
</td>
</tr>
<tr class="row-color" id="i7">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;java.util.ArrayList&lt;java.lang.Double&gt;&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getSidechainCentersOfMass(java.util.HashMap)">getSidechainCentersOfMass</a></span>&#8203;(java.util.HashMap&lt;java.lang.String,&#8203;org.biojava.nbio.structure.Group&gt;&nbsp;residues)</code></th>
<td class="col-last">
<div class="block">Computes the side-chains center of mass for a specified set of amino-acid residues represented as <code>Group</code>
 objects.</div>
</td>
</tr>
<tr class="alt-color" id="i8">
<td class="col-first"><code>private static org.javatuples.Triplet&lt;java.lang.Integer,&#8203;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#globallyMatchNucleotideToAminoAcidSequence(java.lang.String,java.lang.String,java.util.ArrayList,java.lang.String,java.lang.String)">globallyMatchNucleotideToAminoAcidSequence</a></span>&#8203;(java.lang.String&nbsp;aaSeq,
java.lang.String&nbsp;transAASeq,
java.util.ArrayList&lt;java.lang.String&gt;&nbsp;splitNucleotideSequence,
java.lang.String&nbsp;truncatedStart,
java.lang.String&nbsp;truncatedEnd)</code></th>
<td class="col-last">
<div class="block">Matches a nucleotide sequence with an amino-acid sequence based on a global alignment of the amino acid sequence
 with the translated nucleotide sequence.</div>
</td>
</tr>
<tr class="row-color" id="i9">
<td class="col-first"><code>static java.lang.Character</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#invertBase(char)">invertBase</a></span>&#8203;(char&nbsp;base)</code></th>
<td class="col-last">
<div class="block">Returns the complement of a nucleotide base, i.e. switches A to T, C to G and vice versa.</div>
</td>
</tr>
<tr class="alt-color" id="i10">
<td class="col-first"><code>static org.json.simple.JSONObject</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#matchDataToStructure(datastructure.VariantPositionsTable,java.lang.String,org.biojava.nbio.structure.Structure,java.lang.String,java.lang.String,boolean)">matchDataToStructure</a></span>&#8203;(<a href="../datastructure/VariantPositionsTable.html" title="class in datastructure">VariantPositionsTable</a>&nbsp;variantsTable,
java.lang.String&nbsp;referenceAnalysisId,
org.biojava.nbio.structure.Structure&nbsp;proteinStructure,
java.lang.String&nbsp;chainId,
java.lang.String&nbsp;chainAsymId,
boolean&nbsp;isSense)</code></th>
<td class="col-last">
<div class="block">Allocates sample variants data and nucleotide reference data to a specified reference protein structures chain
 positions.</div>
</td>
</tr>
<tr class="row-color" id="i11">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#removeGapsFromSequence(java.lang.String)">removeGapsFromSequence</a></span>&#8203;(java.lang.String&nbsp;sequence)</code></th>
<td class="col-last">
<div class="block">Removes all gap symbols ("-") from the passed sequence.</div>
</td>
</tr>
<tr class="alt-color" id="i12">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#translateCodon(java.lang.String,boolean,boolean)">translateCodon</a></span>&#8203;(java.lang.String&nbsp;codon,
boolean&nbsp;asAA3,
boolean&nbsp;includeTermination)</code></th>
<td class="col-last">
<div class="block">Returns a <code>String</code> representing the one-letter code of the amino-acid that is translated from the passed
 nucleotide codon.</div>
</td>
</tr>
<tr class="row-color" id="i13">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#translateNucSequence(java.util.ArrayList)">translateNucSequence</a></span>&#8203;(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;splitNucSequence)</code></th>
<td class="col-last">
<div class="block">Translates a nucleotide sequence, split into codons of length three, into a single-letter amino acid sequence.</div>
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="inherited-list">
<h3 id="methods.inherited.from.class.java.lang.Object">Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></div>
</section>
</li>
</ul>
</section>
<section class="details">
<ul class="details-list">
<!-- ============ FIELD DETAIL =========== -->
<li>
<section class="field-details" id="field.detail">
<h2>Field Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="NO_MATCH_CHAR">
<h3>NO_MATCH_CHAR</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">NO_MATCH_CHAR</span></div>
<div class="block">Character to indicate a gap/missing information regarding a nucleotide and amino acid sequence.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.NO_MATCH_CHAR">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="TRUNCATED_CHAR">
<h3>TRUNCATED_CHAR</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">TRUNCATED_CHAR</span></div>
<div class="block">Character to indicate a truncation of a sequence.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.TRUNCATED_CHAR">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="SEPARATOR_CHAR">
<h3>SEPARATOR_CHAR</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">SEPARATOR_CHAR</span></div>
<div class="block">Character used as a separator in padded sequences.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.SEPARATOR_CHAR">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="TERMINATION_AA1">
<h3>TERMINATION_AA1</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">TERMINATION_AA1</span></div>
<div class="block">One letter code character used to indicate a translated stop codon.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.TERMINATION_AA1">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="TERMINATION_AA3">
<h3>TERMINATION_AA3</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">TERMINATION_AA3</span></div>
<div class="block">Three letter code String used to indicate a translated stop codon.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.TERMINATION_AA3">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="UNKNOWN_AA1">
<h3>UNKNOWN_AA1</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">UNKNOWN_AA1</span></div>
<div class="block">One letter code character used to indicate the translation of a codon containing an N character (i.e. unknown or
 any nucleotide).</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.UNKNOWN_AA1">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="UNKNOWN_AA3">
<h3>UNKNOWN_AA3</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">UNKNOWN_AA3</span></div>
<div class="block">Three letter code string used to indicate the translation of a codon containing an N character (i.e. unknown or
 any nucleotide).</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.UNKNOWN_AA3">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="AMBIGUOUS_AA3">
<h3>AMBIGUOUS_AA3</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">AMBIGUOUS_AA3</span></div>
<div class="block">Three letter code string used to indicate an ambiguous translation of a codon (i.e. due to an translational
 shift induced by indels).</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.AMBIGUOUS_AA3">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="CODON_MAP">
<h3>CODON_MAP</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">CODON_MAP</span></div>
<div class="block">Hash map mapping nucleotide codons to single-letter amino acids. Stop codons map to no character.</div>
</section>
</li>
<li>
<section class="detail" id="AA1TO3">
<h3>AA1TO3</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">AA1TO3</span></div>
<div class="block">Hash map mapping amino-acid one to three letter codes.</div>
</section>
</li>
<li>
<section class="detail" id="AA1_BLOSUM80_INDEX">
<h3>AA1_BLOSUM80_INDEX</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.Character,&#8203;java.lang.Integer&gt;</span>&nbsp;<span class="member-name">AA1_BLOSUM80_INDEX</span></div>
<div class="block">Hash map mapping amino-acid one letter symbols to indices used to access substitution matrix values.</div>
</section>
</li>
<li>
<section class="detail" id="BLOSUM80">
<h3>BLOSUM80</h3>
<div class="member-signature"><span class="modifiers">private static final</span>&nbsp;<span class="return-type">int[][]</span>&nbsp;<span class="member-name">BLOSUM80</span></div>
<div class="block">The BLOSUM80 matrix used for alignment computation.
 <p>
 Includes values for 'X'; any amino-acid.</div>
</section>
</li>
<li>
<section class="detail" id="GAP_OPEN_PENALTY">
<h3>GAP_OPEN_PENALTY</h3>
<div class="member-signature"><span class="modifiers">private static final</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="member-name">GAP_OPEN_PENALTY</span></div>
<div class="block">Gap opening penalty used for alignment computation.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.GAP_OPEN_PENALTY">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="GAP_EXTENSION_PENALTY">
<h3>GAP_EXTENSION_PENALTY</h3>
<div class="member-signature"><span class="modifiers">private static final</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="member-name">GAP_EXTENSION_PENALTY</span></div>
<div class="block">Gap extension penalty used for alignment computation.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.GAP_EXTENSION_PENALTY">Constant Field Values</a></dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<li>
<section class="constructor-details" id="constructor.detail">
<h2>Constructor Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="&lt;init&gt;()">
<h3>Bio</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="member-name">Bio</span>()</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ============ METHOD DETAIL ========== -->
<li>
<section class="method-details" id="method.detail">
<h2>Method Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="getSequenceFromTable(datastructure.VariantPositionsTable,java.lang.String,java.lang.String,java.util.Iterator,boolean,boolean)">
<h3>getSequenceFromTable</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">getSequenceFromTable</span>&#8203;(<span class="parameters"><a href="../datastructure/VariantPositionsTable.html" title="class in datastructure">VariantPositionsTable</a>&nbsp;table,
java.lang.String&nbsp;referenceAnalysisId,
java.lang.String&nbsp;sampleName,
java.util.Iterator&lt;java.lang.String&gt;&nbsp;positions,
boolean&nbsp;addDeletions,
boolean&nbsp;addInsertions)</span></div>
<div class="block">Constructs and returns a nucleotide sequence as <code>String</code> for the specified sample by using the specified
 variants table,
 reference feature and positions.
 <p>
 The boolean parameters `addDeletions` and `addInsertions` indicate whether insertion and deletion information of
 the sample should be included into the constructed sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>table</code> - <a href="../datastructure/VariantPositionsTable.html" title="class in datastructure"><code>VariantPositionsTable</code></a> from which variant contents are retrieved.</dd>
<dd><code>referenceAnalysisId</code> - <code>String</code> specifying the reference feature to which regard the sequence is
                            constructed.</dd>
<dd><code>sampleName</code> - <code>String</code> specifying the sample which variant contents should be incoorporated into the
                            sequence.</dd>
<dd><code>positions</code> - <code>java.util.Iterator&lt;java.lang.String&gt;</code> specifying which positions are used to construct the sequence; this
                            should be `table.getAllPositionsIterator(referenceAnalysisId)` in order
                            to obtain a biologically meaningful result.</dd>
<dd><code>addDeletions</code> - <code>Boolean</code> whether deletions of the specified sample should be considered, i.e. if the
                            sample shows a deletion on a specific position this position is skipped for
                            constructing the sequence.</dd>
<dd><code>addInsertions</code> - <code>Boolean</code> whether insertions of the specified sample should be considered.</dd>
<dt>Returns:</dt>
<dd><code>String</code> representing the reconstructed nucleotide sequence.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getInsertionsOfSample(datastructure.VariantPositionsTable,java.lang.String,java.lang.String,java.util.NavigableSet,boolean)">
<h3>getInsertionsOfSample</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.Integer,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">getInsertionsOfSample</span>&#8203;(<span class="parameters"><a href="../datastructure/VariantPositionsTable.html" title="class in datastructure">VariantPositionsTable</a>&nbsp;table,
java.lang.String&nbsp;referenceAnalysisId,
java.lang.String&nbsp;sampleName,
java.util.NavigableSet&lt;java.lang.String&gt;&nbsp;positions,
boolean&nbsp;asSense)</span>
                                                                                   throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Constructs and returns a <code>HashMap</code> mapping <code>String</code> representations of genome positions to
 <code>String</code>s representing inserted nucleotide sequences with regard to one analyzed sample.
 <p>
 The positions are thereby chosen as the position after which the respective insertion is observed.
 <p>
 FIXME: In the case of sense orientation and insertions at the very start and anti-sense orientation and
 insertions at the very end, no position can be selected to assign the insertion to.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>table</code> - <a href="../datastructure/VariantPositionsTable.html" title="class in datastructure"><code>VariantPositionsTable</code></a> from which variant contents are retrieved.</dd>
<dd><code>referenceAnalysisId</code> - <code>String</code> specifying the reference feature to which regard the insertions are
                            extracted.</dd>
<dd><code>sampleName</code> - <code>String</code> specifying the sample for which insertions shall be extracted.</dd>
<dd><code>positions</code> - <code>java.util.Iterator&lt;java.lang.String&gt;</code> specifying which positions are scanned for insertions; this
                            should be `table.getAllPositionsIterator(referenceAnalysisId)` in order
                            to obtain a biologically meaningful result.</dd>
<dd><code>asSense</code> - <code>Boolean</code> specifying whether the insertions should be extracted with respect to the sense
                            or anti-sense strand; i.e. if set to false the stored insertions are the reverse
                            complement sequences of the determined insertions and the position is set to -
                            (i+1) if i is the position of the insertion in sense orientation.</dd>
<dt>Returns:</dt>
<dd><code>HashMap</code> mapping genome positions to inserted nucleotide sequences.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If, during reverse complement computation in the case of anti-sense representation,
                            any base can not be translated.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getNucleotideToAminoAcidSequenceMatching(java.lang.String,java.lang.String)">
<h3>getNucleotideToAminoAcidSequenceMatching</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">org.javatuples.Pair&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">getNucleotideToAminoAcidSequenceMatching</span>&#8203;(<span class="parameters">java.lang.String&nbsp;aminoAcidSequence,
java.lang.String&nbsp;nucleotideSequence)</span>
                                                                                                       throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Matches a amino-acid and nucleotide sequence.
 <p>
 This is done by translating the nucleotide sequence, for every possible frameshift if applicable, into an amino
 acid sequence and aligning the translated nucleotide sequence with the amino-acid sequence globally.
 <p>
 The alignment is used to deduce a mapping of nucleotide codons to the amino acid sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>aminoAcidSequence</code> - <code>String</code> representing a amino-acid sequence.</dd>
<dd><code>nucleotideSequence</code> - <code>String</code> representing a nucleotide sequence.</dd>
<dt>Returns:</dt>
<dd><code>java.util.ArrayList&lt;java.lang.String&gt;</code> comprising the padded amino acid (at index 0) and nucleotide sequence (at
 index 1) that were derived as the best alignment.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If an error occurs during the translation of the nucleotide sequence, i.e. any codon
                            with a length different than three is detected or any codon can not be translated.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="globallyMatchNucleotideToAminoAcidSequence(java.lang.String,java.lang.String,java.util.ArrayList,java.lang.String,java.lang.String)">
<h3>globallyMatchNucleotideToAminoAcidSequence</h3>
<div class="member-signature"><span class="modifiers">private static</span>&nbsp;<span class="return-type">org.javatuples.Triplet&lt;java.lang.Integer,&#8203;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">globallyMatchNucleotideToAminoAcidSequence</span>&#8203;(<span class="parameters">java.lang.String&nbsp;aaSeq,
java.lang.String&nbsp;transAASeq,
java.util.ArrayList&lt;java.lang.String&gt;&nbsp;splitNucleotideSequence,
java.lang.String&nbsp;truncatedStart,
java.lang.String&nbsp;truncatedEnd)</span></div>
<div class="block">Matches a nucleotide sequence with an amino-acid sequence based on a global alignment of the amino acid sequence
 with the translated nucleotide sequence.
 <p>
 The global gap-affine alignment computation uses the BLOSUM80 scoring matrix, can handle any amino-acid
 symbolized by an 'X' and uses an gap-opening and extension penalty of 6 and 4, respectively.
 <p>
 The aligned translated nucleotide sequence is mapped back to a nucleotide sequence and possible truncated
 nucleotides are re-appended. The aligned amino-acid sequence is represented as three-letter code amino-acids
 separated by a separator symbol. The aligned nucleotide sequence is represented by blocks of length three, i.e.
 codons, separated by a separator symbol. Instead of the common alignment gap symbol '-', the symbol '_' is used
 as '-' is internally reserved for deletions wrt. nucleotide sequences.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>aaSeq</code> - <code>String</code> representing a non-nucleotide-derived amino acid sequence.</dd>
<dd><code>transAASeq</code> - <code>String</code> representing a amino acid sequence that originates from translating a
                                nucleotide sequence.</dd>
<dd><code>splitNucleotideSequence</code> - <code>String</code> the original nucleotide sequence split into chunks of length three, i.e.
                                codons.</dd>
<dd><code>truncatedStart</code> - <code>String</code> nucleotides that were truncated from the start of the nucleotide sequence
                                due to a frameshift.</dd>
<dd><code>truncatedEnd</code> - <code>String</code> nucleotides that were truncated from the end of the nucleotide sequence due
                                to a frameshift.</dd>
<dt>Returns:</dt>
<dd><code>Triplet</code> yielding the alignment score in the first field, the matched amino-acid sequence in the
 second field and the matched nucleotide sequence in the third field.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="matchDataToStructure(datastructure.VariantPositionsTable,java.lang.String,org.biojava.nbio.structure.Structure,java.lang.String,java.lang.String,boolean)">
<h3>matchDataToStructure</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">org.json.simple.JSONObject</span>&nbsp;<span class="member-name">matchDataToStructure</span>&#8203;(<span class="parameters"><a href="../datastructure/VariantPositionsTable.html" title="class in datastructure">VariantPositionsTable</a>&nbsp;variantsTable,
java.lang.String&nbsp;referenceAnalysisId,
org.biojava.nbio.structure.Structure&nbsp;proteinStructure,
java.lang.String&nbsp;chainId,
java.lang.String&nbsp;chainAsymId,
boolean&nbsp;isSense)</span>
                                                       throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Allocates sample variants data and nucleotide reference data to a specified reference protein structures chain
 positions.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>variantsTable</code> - <a href="../datastructure/VariantPositionsTable.html" title="class in datastructure"><code>VariantPositionsTable</code></a> instance yielding all sample variants information.</dd>
<dd><code>referenceAnalysisId</code> - <code>String</code> specifying the reference feature that shall be allocated to the
                            protein structure.</dd>
<dd><code>proteinStructure</code> - <code>Structure</code> yielding the protein structure that shall be used as the with the
                            reference feature associated protein structure.</dd>
<dd><code>chainId</code> - <code>String</code> chain identifier to access protein data from <code>Structure</code> objects and naming.</dd>
<dd><code>chainAsymId</code> - <code>String</code> internal chain identifier to access protein data from <code>Structure</code> objects.</dd>
<dd><code>isSense</code> - <code>Boolean</code> indicating whether the feature to analyze is located on the sense strand.</dd>
<dt>Returns:</dt>
<dd><code>JSONObject</code> yielding sample and reference nucleotide information allocated to protein positions.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If any invoked method call (i.e. codon translation or base reversal) throws this
                            exception.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="translateNucSequence(java.util.ArrayList)">
<h3>translateNucSequence</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">translateNucSequence</span>&#8203;(<span class="parameters">java.util.ArrayList&lt;java.lang.String&gt;&nbsp;splitNucSequence)</span>
                                             throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Translates a nucleotide sequence, split into codons of length three, into a single-letter amino acid sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>splitNucSequence</code> - <code>java.util.ArrayList&lt;java.lang.String&gt;</code> representing a nucleotide sequence that was split into codons
                         of length 3.</dd>
<dt>Returns:</dt>
<dd><code>String</code> representing the translated nucleotide sequence.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If any codon with a length different than three is detected.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getReverseComplement(java.lang.String)">
<h3>getReverseComplement</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">getReverseComplement</span>&#8203;(<span class="parameters">java.lang.String&nbsp;sequence)</span>
                                             throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Returns the reverse complement of the passed nucleotide sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>sequence</code> - <code>String</code> representing a nucleotide sequence.</dd>
<dt>Returns:</dt>
<dd><code>String</code> representing the reverse complement of the passed sequence.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If any base occurs in the sequence that can not be reversed, i.e. for which no
                            complement base is known.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="invertBase(char)">
<h3>invertBase</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.Character</span>&nbsp;<span class="member-name">invertBase</span>&#8203;(<span class="parameters">char&nbsp;base)</span>
                                      throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Returns the complement of a nucleotide base, i.e. switches A to T, C to G and vice versa. For non nucleotide
 symbols, e.g. gaps (-) or any nucleotides (N), the identity is returned.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>base</code> - <code>Character</code> base to invert.</dd>
<dt>Returns:</dt>
<dd><code>Character</code> the inverted base.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If no complement base for the specified base is known.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="translateCodon(java.lang.String,boolean,boolean)">
<h3>translateCodon</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">translateCodon</span>&#8203;(<span class="parameters">java.lang.String&nbsp;codon,
boolean&nbsp;asAA3,
boolean&nbsp;includeTermination)</span>
                                       throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Returns a <code>String</code> representing the one-letter code of the amino-acid that is translated from the passed
 nucleotide codon.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>codon</code> - <code>String</code> representing the nucleotide codon to translate.</dd>
<dd><code>asAA3</code> - <code>Boolean</code> whether the codon should be translated to the amino-acid three-letter code or not.</dd>
<dd><code>includeTermination</code> - <code>Boolean</code> whether a representation for translation termination (in the case of
                           three-letter code 'TER' and in the case of one-letter code 'Z') shall
                           be returned or not, i.e. an empty String is returned instead.</dd>
<dt>Returns:</dt>
<dd><code>String</code> representing a translated amino-acid.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If the passed codon has a length other than three.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="removeGapsFromSequence(java.lang.String)">
<h3>removeGapsFromSequence</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">removeGapsFromSequence</span>&#8203;(<span class="parameters">java.lang.String&nbsp;sequence)</span></div>
<div class="block">Removes all gap symbols ("-") from the passed sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>sequence</code> - The sequence from which gaps should be removed.</dd>
<dt>Returns:</dt>
<dd><code>String</code>, the passed sequence without any gap symbols.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getSequencesFromPdb(org.biojava.nbio.structure.Structure)">
<h3>getSequencesFromPdb</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">getSequencesFromPdb</span>&#8203;(<span class="parameters">org.biojava.nbio.structure.Structure&nbsp;structure)</span></div>
<div class="block">Returns the chains sequences from a <code>Structure</code> instance parsed from a .pdb file.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>structure</code> - <code>Structure</code> instance to extract the chains sequences from.</dd>
<dt>Returns:</dt>
<dd><code>HashMap</code> storing the .pdb files amino-acid sequences per chain.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getFirstResidueNumberFromProteinStructure(org.biojava.nbio.structure.Structure,java.lang.String)">
<h3>getFirstResidueNumberFromProteinStructure</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="member-name">getFirstResidueNumberFromProteinStructure</span>&#8203;(<span class="parameters">org.biojava.nbio.structure.Structure&nbsp;structure,
java.lang.String&nbsp;chainAsymId)</span></div>
<div class="block">Returns the first residue number from a <code>Structure</code> instances specified chain (via chainId).</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>structure</code> - <code>Structure</code> instance to access.</dd>
<dd><code>chainAsymId</code> - <code>String</code> representing the AsymId of the chain to access.</dd>
<dt>Returns:</dt>
<dd>The first residue number of the specified structure and chain.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getResiduesFromStructure(org.biojava.nbio.structure.Structure)">
<h3>getResiduesFromStructure</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;org.biojava.nbio.structure.Group&gt;</span>&nbsp;<span class="member-name">getResiduesFromStructure</span>&#8203;(<span class="parameters">org.biojava.nbio.structure.Structure&nbsp;structure)</span></div>
<div class="block">Returns a <code>HashMap</code> mapping residue number and chain names as (concatenated) keys to the respective
 <code>Group</code> instances.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>structure</code> - <code>Structure</code> instance to extract the residue <code>Group</code> objects from.</dd>
<dt>Returns:</dt>
<dd>The mapping of residue identifiers to residues.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getSidechainCentersOfMass(java.util.HashMap)">
<h3>getSidechainCentersOfMass</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.util.ArrayList&lt;java.lang.Double&gt;&gt;</span>&nbsp;<span class="member-name">getSidechainCentersOfMass</span>&#8203;(<span class="parameters">java.util.HashMap&lt;java.lang.String,&#8203;org.biojava.nbio.structure.Group&gt;&nbsp;residues)</span></div>
<div class="block">Computes the side-chains center of mass for a specified set of amino-acid residues represented as <code>Group</code>
 objects.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>residues</code> - <code>HashMap</code> mapping from residue identifiers to <code>Group</code> instances.</dd>
<dt>Returns:</dt>
<dd>The mapping of residue identifiers to residue side-chains center of mass.</dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
</ul>
</section>
<!-- ========= END OF CLASS DATA ========= -->
</main>
<footer role="contentinfo">
<nav role="navigation">
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottom-nav" id="navbar.bottom">
<div class="skip-nav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.bottom.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
</div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<span class="skip-nav" id="skip.navbar.bottom">
<!--   -->
</span></nav>
</footer>
</div>
</div>
</body>
</html>
