<!DOCTYPE HTML>
<!-- NewPage -->
<html lang="de">
<head>
<!-- Generated by javadoc (15) on Thu Nov 11 10:13:35 CET 2021 -->
<title>Bio</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="dc.created" content="2021-11-11">
<meta name="description" content="declaration: package: utility, class: Bio">
<meta name="generator" content="javadoc/ClassWriterImpl">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
<script type="text/javascript" src="../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../script-dir/jquery-ui.min.js"></script>
</head>
<body class="class-declaration-page">
<script type="text/javascript">var data = {"i0":9,"i1":9,"i2":9,"i3":9,"i4":9,"i5":9,"i6":9,"i7":9,"i8":9,"i9":9,"i10":9};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],8:["t4","Concrete Methods"]};
var altColor = "alt-color";
var rowColor = "row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar.top">
<div class="skip-nav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.top.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search">SEARCH:</label>
<input type="text" id="search" value="search" disabled="disabled">
<input type="reset" id="reset" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip.navbar.top">
<!--   -->
</span></nav>
</header>
<div class="flex-content">
<main role="main">
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="sub-title"><span class="package-label-in-type">Package</span>&nbsp;<a href="package-summary.html">utility</a></div>
<h1 title="Class Bio" class="title">Class Bio</h1>
</div>
<div class="inheritance" title="Inheritance Tree">java.lang.Object
<div class="inheritance">utility.Bio</div>
</div>
<section class="description">
<hr>
<pre>public final class <span class="type-name-label">Bio</span>
extends java.lang.Object</pre>
<div class="block">Comprises static methods that implement algorithms to solve specifically bioinformatics problems.
 <p>
 Currently only methods for sequence alignment exist which are used to align a nucleotide with a amino acid
 sequence. This is done in a special way by first translating the nucleotide sequence into an amino acid sequence.
 The alignment information is used to transform the translated nucleotide sequence back into a nucleotide sequence.
 This process is applied enable a visualization of the respective sequences and should not be used for general
 local sequence alignment.</div>
<dl class="notes">
<dt>Since:</dt>
<dd>2.0</dd>
<dt>Version:</dt>
<dd>2.0</dd>
<dt>Author:</dt>
<dd>Simon Hackl</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<!-- =========== FIELD SUMMARY =========== -->
<li>
<section class="field-summary" id="field.summary">
<h2>Field Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Fields</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Field</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#AA1TO3">AA1TO3</a></span></code></th>
<td class="col-last">
<div class="block">Hash map mapping amino-acid one to three letter codes.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>private static org.biojava.nbio.core.alignment.template.SubstitutionMatrix&lt;org.biojava.nbio.core.sequence.compound.AminoAcidCompound&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#BLOSUM62">BLOSUM62</a></span></code></th>
<td class="col-last">
<div class="block">TODO: Use other or own substitution matrix, especially to consider internal stop codons, if present.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#CODON_MAP">CODON_MAP</a></span></code></th>
<td class="col-last">
<div class="block">Hash map mapping nucleotide codons to single-letter amino acids.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#NO_MATCH_CHAR">NO_MATCH_CHAR</a></span></code></th>
<td class="col-last">
<div class="block">Character to indicate a gap/missing information regarding a nucleotide and amino acid sequence.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#SEPARATOR_CHAR">SEPARATOR_CHAR</a></span></code></th>
<td class="col-last">
<div class="block">Character used as a separator in padded sequences.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#STOP_CODON_CHAR">STOP_CODON_CHAR</a></span></code></th>
<td class="col-last">
<div class="block">Character to indicate stop codon.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>static char</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#TRUNCATED_CHAR">TRUNCATED_CHAR</a></span></code></th>
<td class="col-last">
<div class="block">Character to indicate a truncation of a sequence.</div>
</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<li>
<section class="constructor-summary" id="constructor.summary">
<h2>Constructor Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Constructors</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Constructor</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<th class="col-constructor-name" scope="row"><code><span class="member-name-link"><a href="#%3Cinit%3E()">Bio</a></span>()</code></th>
<td class="col-last">&nbsp;</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ========== METHOD SUMMARY =========== -->
<li>
<section class="method-summary" id="method.summary">
<h2>Method Summary</h2>
<div class="member-summary" id="method-summary-table">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button role="tab" aria-selected="true" aria-controls="method-summary-table.tabpanel" tabindex="0" onkeydown="switchTab(event)" id="t0" class="active-table-tab">All Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t1" class="table-tab" onclick="show(1);">Static Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t4" class="table-tab" onclick="show(8);">Concrete Methods</button></div>
<div id="method-summary-table.tabpanel" role="tabpanel">
<table class="summary-table" aria-labelledby="t0">
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Method</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color" id="i0">
<td class="col-first"><code>static int</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getFirstResidueNumberFromProteinStructure(org.biojava.nbio.structure.Structure,java.lang.String)">getFirstResidueNumberFromProteinStructure</a></span>&#8203;(org.biojava.nbio.structure.Structure&nbsp;structure,
java.lang.String&nbsp;chainAsymId)</code></th>
<td class="col-last">
<div class="block">Returns the first residue number from a <code>Structure</code> instances specified chain (via chainId).</div>
</td>
</tr>
<tr class="row-color" id="i1">
<td class="col-first"><code>static int</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getPositionOnReverseComplement(int,int)">getPositionOnReverseComplement</a></span>&#8203;(int&nbsp;position,
int&nbsp;sequenceLength)</code></th>
<td class="col-last">
<div class="block">Returns the position q of a position p on the reverse complement of a genomic feature.</div>
</td>
</tr>
<tr class="alt-color" id="i2">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;org.biojava.nbio.structure.Group&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getResiduesFromStructure(org.biojava.nbio.structure.Structure)">getResiduesFromStructure</a></span>&#8203;(org.biojava.nbio.structure.Structure&nbsp;structure)</code></th>
<td class="col-last">
<div class="block">Returns a <code>HashMap</code> mapping residue number and chain names as (concatenated) keys to the respective
 <code>Group</code> instances.</div>
</td>
</tr>
<tr class="row-color" id="i3">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getReverseComplement(java.lang.String)">getReverseComplement</a></span>&#8203;(java.lang.String&nbsp;sequence)</code></th>
<td class="col-last">
<div class="block">Returns the reverse complement of the passed nucleotide sequence.</div>
</td>
</tr>
<tr class="alt-color" id="i4">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getSequencesFromPdb(org.biojava.nbio.structure.Structure)">getSequencesFromPdb</a></span>&#8203;(org.biojava.nbio.structure.Structure&nbsp;structure)</code></th>
<td class="col-last">
<div class="block">Returns the chains sequences from a <code>Structure</code> instance parsed from a .pdb file.</div>
</td>
</tr>
<tr class="row-color" id="i5">
<td class="col-first"><code>static java.util.HashMap&lt;java.lang.String,&#8203;java.util.ArrayList&lt;java.lang.Double&gt;&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getSidechainCentersOfMass(java.util.HashMap)">getSidechainCentersOfMass</a></span>&#8203;(java.util.HashMap&lt;java.lang.String,&#8203;org.biojava.nbio.structure.Group&gt;&nbsp;residues)</code></th>
<td class="col-last">
<div class="block">Computes the side-chains center of mass for a specified set of amino-acid residues represented as <code>Group</code>
 objects.</div>
</td>
</tr>
<tr class="alt-color" id="i6">
<td class="col-first"><code>static java.lang.Character</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#invertBase(char)">invertBase</a></span>&#8203;(char&nbsp;base)</code></th>
<td class="col-last">
<div class="block">Returns the complement of a nucleotide base, i.e. switches A to T, C to G and vice versa.</div>
</td>
</tr>
<tr class="row-color" id="i7">
<td class="col-first"><code>static java.util.ArrayList&lt;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#matchSequences(java.lang.String,java.lang.String)">matchSequences</a></span>&#8203;(java.lang.String&nbsp;aaSequence,
java.lang.String&nbsp;nucSequence)</code></th>
<td class="col-last">
<div class="block">Matches a amino-acid and nucleotide sequence.</div>
</td>
</tr>
<tr class="alt-color" id="i8">
<td class="col-first"><code>private static void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#padSequences(java.lang.String,java.lang.String,java.util.ArrayList,java.lang.String,java.lang.String,java.util.HashMap)">padSequences</a></span>&#8203;(java.lang.String&nbsp;aaSequence,
java.lang.String&nbsp;translatedAASequence,
java.util.ArrayList&lt;java.lang.String&gt;&nbsp;splitNucleotideSequence,
java.lang.String&nbsp;truncatedStart,
java.lang.String&nbsp;truncatedEnd,
java.util.HashMap&lt;java.lang.Integer,&#8203;java.util.ArrayList&lt;java.lang.String&gt;&gt;&nbsp;resultsMap)</code></th>
<td class="col-last">
<div class="block">Conducts a local alignment of amino acid sequences using the BLOSUM62 scoring matrix.</div>
</td>
</tr>
<tr class="row-color" id="i9">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#removeGapsFromSequence(java.lang.String)">removeGapsFromSequence</a></span>&#8203;(java.lang.String&nbsp;sequence)</code></th>
<td class="col-last">
<div class="block">Removes all gap symbols ("-") from the passed sequence.</div>
</td>
</tr>
<tr class="alt-color" id="i10">
<td class="col-first"><code>private static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#translateNucSequence(java.util.ArrayList)">translateNucSequence</a></span>&#8203;(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;splitNucSequence)</code></th>
<td class="col-last">
<div class="block">Translates a nucleotide sequence, split into codons of length three, into a single-letter amino acid sequence.</div>
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="inherited-list">
<h3 id="methods.inherited.from.class.java.lang.Object">Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></div>
</section>
</li>
</ul>
</section>
<section class="details">
<ul class="details-list">
<!-- ============ FIELD DETAIL =========== -->
<li>
<section class="field-details" id="field.detail">
<h2>Field Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="NO_MATCH_CHAR">
<h3>NO_MATCH_CHAR</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">NO_MATCH_CHAR</span></div>
<div class="block">Character to indicate a gap/missing information regarding a nucleotide and amino acid sequence.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.NO_MATCH_CHAR">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="TRUNCATED_CHAR">
<h3>TRUNCATED_CHAR</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">TRUNCATED_CHAR</span></div>
<div class="block">Character to indicate a truncation of a sequence.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.TRUNCATED_CHAR">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="STOP_CODON_CHAR">
<h3>STOP_CODON_CHAR</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">STOP_CODON_CHAR</span></div>
<div class="block">Character to indicate stop codon.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.STOP_CODON_CHAR">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="SEPARATOR_CHAR">
<h3>SEPARATOR_CHAR</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">char</span>&nbsp;<span class="member-name">SEPARATOR_CHAR</span></div>
<div class="block">Character used as a separator in padded sequences.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#utility.Bio.SEPARATOR_CHAR">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="CODON_MAP">
<h3>CODON_MAP</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">CODON_MAP</span></div>
<div class="block">Hash map mapping nucleotide codons to single-letter amino acids. Stop codons map to no character.</div>
</section>
</li>
<li>
<section class="detail" id="AA1TO3">
<h3>AA1TO3</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">AA1TO3</span></div>
<div class="block">Hash map mapping amino-acid one to three letter codes.</div>
</section>
</li>
<li>
<section class="detail" id="BLOSUM62">
<h3>BLOSUM62</h3>
<div class="member-signature"><span class="modifiers">private static final</span>&nbsp;<span class="return-type">org.biojava.nbio.core.alignment.template.SubstitutionMatrix&lt;org.biojava.nbio.core.sequence.compound.AminoAcidCompound&gt;</span>&nbsp;<span class="member-name">BLOSUM62</span></div>
<div class="block">TODO: Use other or own substitution matrix, especially to consider internal stop codons, if present.
 The BLOSUM62 substitution matrix.</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<li>
<section class="constructor-details" id="constructor.detail">
<h2>Constructor Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="&lt;init&gt;()">
<h3>Bio</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="member-name">Bio</span>()</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ============ METHOD DETAIL ========== -->
<li>
<section class="method-details" id="method.detail">
<h2>Method Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="matchSequences(java.lang.String,java.lang.String)">
<h3>matchSequences</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.ArrayList&lt;java.lang.String&gt;</span>&nbsp;<span class="member-name">matchSequences</span>&#8203;(<span class="parameters">java.lang.String&nbsp;aaSequence,
java.lang.String&nbsp;nucSequence)</span>
                                                            throws <span class="exceptions">org.biojava.nbio.core.exceptions.CompoundNotFoundException</span></div>
<div class="block">Matches a amino-acid and nucleotide sequence.
 <p>
 This is done by translating the nucleotide sequence, for every possible frameshift if applicable, into a amino
 acid sequence and aligning the translated nucleotide sequence with the amino-acid sequence. The best alignment,
 here the one with the least number of gaps, is used as the best matching sequences.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>aaSequence</code> - <code>String</code> representing a amino-acid sequence.</dd>
<dd><code>nucSequence</code> - <code>String</code> representing a nucleotide sequence.</dd>
<dt>Returns:</dt>
<dd><code>java.util.ArrayList&lt;java.lang.String&gt;</code> comprising the padded amino acid (at index 0) and nucleotide sequence (at
 index 1) that were derived as the best alignment.</dd>
<dt>Throws:</dt>
<dd><code>org.biojava.nbio.core.exceptions.CompoundNotFoundException</code> - If a compound, i.e. amino acid, could not be translated.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="translateNucSequence(java.util.ArrayList)">
<h3>translateNucSequence</h3>
<div class="member-signature"><span class="modifiers">private static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">translateNucSequence</span>&#8203;(<span class="parameters">java.util.ArrayList&lt;java.lang.String&gt;&nbsp;splitNucSequence)</span></div>
<div class="block">Translates a nucleotide sequence, split into codons of length three, into a single-letter amino acid sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>splitNucSequence</code> - <code>java.util.ArrayList&lt;java.lang.String&gt;</code> representing a nucleotide sequence that was split into codons
                         of length 3.</dd>
<dt>Returns:</dt>
<dd><code>String</code> representing the translated nucleotide sequence.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="padSequences(java.lang.String,java.lang.String,java.util.ArrayList,java.lang.String,java.lang.String,java.util.HashMap)">
<h3>padSequences</h3>
<div class="member-signature"><span class="modifiers">private static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">padSequences</span>&#8203;(<span class="parameters">java.lang.String&nbsp;aaSequence,
java.lang.String&nbsp;translatedAASequence,
java.util.ArrayList&lt;java.lang.String&gt;&nbsp;splitNucleotideSequence,
java.lang.String&nbsp;truncatedStart,
java.lang.String&nbsp;truncatedEnd,
java.util.HashMap&lt;java.lang.Integer,&#8203;java.util.ArrayList&lt;java.lang.String&gt;&gt;&nbsp;resultsMap)</span>
                          throws <span class="exceptions">org.biojava.nbio.core.exceptions.CompoundNotFoundException</span></div>
<div class="block">Conducts a local alignment of amino acid sequences using the BLOSUM62 scoring matrix.
 <p>
 Thereby, one amino acid sequence is derived from a nucleotide sequence.
 <p>
 Alignment gaps are replaced by special characters to separate alignment gaps from such gaps that are
 introduces to the passed nucleotide sequence by insertions of samples. In addition, the non-nucleotide-derived
 amino acid sequence is padded with symbols to represent a truncation of the original nucleotide sequence that
 may be present from a possible frameshift.
 <p>
 The aligned translated nucleotide sequence is mapped back to a nucleotide sequence and, both, the nucleotide
 sequence and amino acid sequence are stored in a map accessible via the number of gaps in the alignment.
 <p>
 Insertions are currently handled separate and not included into the padded sequences directly.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>aaSequence</code> - <code>String</code> representing a non-nucleotide-derived amino acid sequence.</dd>
<dd><code>translatedAASequence</code> - <code>String</code> representing a amino acid sequence that originates from translating a
                                nucleotide sequence.</dd>
<dd><code>splitNucleotideSequence</code> - <code>String</code> the original nucleotide sequence split into chunks of length three, i.e.
                                codons.</dd>
<dd><code>truncatedStart</code> - <code>String</code> nucleotides that were truncated from the start of the nucleotide sequence
                                due to a frameshift.</dd>
<dd><code>truncatedEnd</code> - <code>String</code> nucleotides that were truncated from the end of the nucleotide sequence due
                                to a frameshift.</dd>
<dd><code>resultsMap</code> - <code>HashMap</code> mapping the alignment score, here the number of gaps, to the aligned sequences.</dd>
<dt>Throws:</dt>
<dd><code>org.biojava.nbio.core.exceptions.CompoundNotFoundException</code> - If a compound, i.e. amino acid, could not be translated.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getReverseComplement(java.lang.String)">
<h3>getReverseComplement</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">getReverseComplement</span>&#8203;(<span class="parameters">java.lang.String&nbsp;sequence)</span></div>
<div class="block">Returns the reverse complement of the passed nucleotide sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>sequence</code> - <code>String</code> representing a nucleotide sequence.</dd>
<dt>Returns:</dt>
<dd><code>String</code> representing the reverse complement of the passed sequence.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="invertBase(char)">
<h3>invertBase</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.Character</span>&nbsp;<span class="member-name">invertBase</span>&#8203;(<span class="parameters">char&nbsp;base)</span></div>
<div class="block">Returns the complement of a nucleotide base, i.e. switches A to T, C to G and vice versa.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>base</code> - <code>Character</code> base to invert.</dd>
<dt>Returns:</dt>
<dd><code>Character</code> the inverted base.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getPositionOnReverseComplement(int,int)">
<h3>getPositionOnReverseComplement</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="member-name">getPositionOnReverseComplement</span>&#8203;(<span class="parameters">int&nbsp;position,
int&nbsp;sequenceLength)</span></div>
<div class="block">Returns the position q of a position p on the reverse complement of a genomic feature.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>position</code> - <code>Integer</code> the position that should be converted.</dd>
<dd><code>sequenceLength</code> - <code>Integer</code> the length of the parent sequence (of the genomic
                       feature) or full contig/chromosome/plasmid.</dd>
<dt>Returns:</dt>
<dd><code>Integer</code> representing the position on the reverse complement strand.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="removeGapsFromSequence(java.lang.String)">
<h3>removeGapsFromSequence</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">removeGapsFromSequence</span>&#8203;(<span class="parameters">java.lang.String&nbsp;sequence)</span></div>
<div class="block">Removes all gap symbols ("-") from the passed sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>sequence</code> - The sequence from which gaps should be removed.</dd>
<dt>Returns:</dt>
<dd><code>String</code>, the passed sequence without any gap symbols.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getSequencesFromPdb(org.biojava.nbio.structure.Structure)">
<h3>getSequencesFromPdb</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">getSequencesFromPdb</span>&#8203;(<span class="parameters">org.biojava.nbio.structure.Structure&nbsp;structure)</span></div>
<div class="block">Returns the chains sequences from a <code>Structure</code> instance parsed from a .pdb file.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>structure</code> - <code>Structure</code> instance to extract the chains sequences from.</dd>
<dt>Returns:</dt>
<dd><code>HashMap</code> storing the .pdb files amino-acid sequences per chain.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getFirstResidueNumberFromProteinStructure(org.biojava.nbio.structure.Structure,java.lang.String)">
<h3>getFirstResidueNumberFromProteinStructure</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="member-name">getFirstResidueNumberFromProteinStructure</span>&#8203;(<span class="parameters">org.biojava.nbio.structure.Structure&nbsp;structure,
java.lang.String&nbsp;chainAsymId)</span></div>
<div class="block">Returns the first residue number from a <code>Structure</code> instances specified chain (via chainId).</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>structure</code> - <code>Structure</code> instance to access.</dd>
<dd><code>chainAsymId</code> - <code>String</code> representing the AsymId of the chain to access.</dd>
<dt>Returns:</dt>
<dd>The first residue number of the specified structure and chain.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getResiduesFromStructure(org.biojava.nbio.structure.Structure)">
<h3>getResiduesFromStructure</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;org.biojava.nbio.structure.Group&gt;</span>&nbsp;<span class="member-name">getResiduesFromStructure</span>&#8203;(<span class="parameters">org.biojava.nbio.structure.Structure&nbsp;structure)</span></div>
<div class="block">Returns a <code>HashMap</code> mapping residue number and chain names as (concatenated) keys to the respective
 <code>Group</code> instances.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>structure</code> - <code>Structure</code> instance to extract the residue <code>Group</code> objects from.</dd>
<dt>Returns:</dt>
<dd>The mapping of residue identifiers to residues.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getSidechainCentersOfMass(java.util.HashMap)">
<h3>getSidechainCentersOfMass</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.util.ArrayList&lt;java.lang.Double&gt;&gt;</span>&nbsp;<span class="member-name">getSidechainCentersOfMass</span>&#8203;(<span class="parameters">java.util.HashMap&lt;java.lang.String,&#8203;org.biojava.nbio.structure.Group&gt;&nbsp;residues)</span></div>
<div class="block">Computes the side-chains center of mass for a specified set of amino-acid residues represented as <code>Group</code>
 objects.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>residues</code> - <code>HashMap</code> mapping from residue identifiers to <code>Group</code> instances.</dd>
<dt>Returns:</dt>
<dd>The mapping of residue identifiers to residue side-chains center of mass.</dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
</ul>
</section>
<!-- ========= END OF CLASS DATA ========= -->
</main>
<footer role="contentinfo">
<nav role="navigation">
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottom-nav" id="navbar.bottom">
<div class="skip-nav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.bottom.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
</div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<span class="skip-nav" id="skip.navbar.bottom">
<!--   -->
</span></nav>
</footer>
</div>
</div>
</body>
</html>
