<!DOCTYPE HTML>
<!-- NewPage -->
<html lang="de">
<head>
<!-- Generated by javadoc (15) on Mon Jul 25 13:17:44 CEST 2022 -->
<title>AllocatedProteinEntry</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="dc.created" content="2022-07-25">
<meta name="description" content="declaration: package: datastructure, class: AllocatedProteinEntry">
<meta name="generator" content="javadoc/ClassWriterImpl">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
<script type="text/javascript" src="../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../script-dir/jquery-ui.min.js"></script>
</head>
<body class="class-declaration-page">
<script type="text/javascript">var data = {"i0":10,"i1":10};
var tabs = {65535:["t0","All Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "alt-color";
var rowColor = "row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar.top">
<div class="skip-nav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.top.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search">SEARCH:</label>
<input type="text" id="search" value="search" disabled="disabled">
<input type="reset" id="reset" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip.navbar.top">
<!--   -->
</span></nav>
</header>
<div class="flex-content">
<main role="main">
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="sub-title"><span class="package-label-in-type">Package</span>&nbsp;<a href="package-summary.html">datastructure</a></div>
<h1 title="Class AllocatedProteinEntry" class="title">Class AllocatedProteinEntry</h1>
</div>
<div class="inheritance" title="Inheritance Tree">java.lang.Object
<div class="inheritance">datastructure.AllocatedProteinEntry</div>
</div>
<section class="description">
<hr>
<pre>public class <span class="type-name-label">AllocatedProteinEntry</span>
extends java.lang.Object</pre>
<div class="block">Internal object representation of a protein that was allocated to a single <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>.</div>
<dl class="notes">
<dt>Since:</dt>
<dd>2.0</dd>
<dt>Version:</dt>
<dd>2.1</dd>
<dt>Author:</dt>
<dd>Simon Hackl</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<!-- =========== FIELD SUMMARY =========== -->
<li>
<section class="field-summary" id="field.summary">
<h2>Field Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Fields</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Field</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<td class="col-first"><code>java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#chainSequences">chainSequences</a></span></code></th>
<td class="col-last">
<div class="block">Aligned reference amino acid sequences, one per protein chain, of the entry.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#name">name</a></span></code></th>
<td class="col-last">
<div class="block"><code>String</code>; The internal name of the protein.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#pdb">pdb</a></span></code></th>
<td class="col-last">
<div class="block"><code>String</code> representation of a `pdb` file containing protein structure information of a coding gene feature.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;<a href="ProteoformEntry.html" title="class in datastructure">ProteoformEntry</a>&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#proteoforms">proteoforms</a></span></code></th>
<td class="col-last">
<div class="block">Map structure mapping <code>String</code>s (representing proteoform names) to <a href="ProteoformEntry.html" title="class in datastructure"><code>ProteoformEntry</code></a>s</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;<a href="AminoacidVariantAnnotationEntry.html" title="class in datastructure">AminoacidVariantAnnotationEntry</a>&gt;&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#variants">variants</a></span></code></th>
<td class="col-last">
<div class="block">Hierarchical map structure that uses <code>String</code>s as keys and point to <a href="AminoacidVariantAnnotationEntry.html" title="class in datastructure"><code>AminoacidVariantAnnotationEntry</code></a>s.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#WILD_TYPE_PROTEOFORM_ID">WILD_TYPE_PROTEOFORM_ID</a></span></code></th>
<td class="col-last">
<div class="block"><code>String</code> identifier used to store the wild type proteoform.</div>
</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<li>
<section class="constructor-summary" id="constructor.summary">
<h2>Constructor Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Constructors</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Constructor</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<th class="col-constructor-name" scope="row"><code><span class="member-name-link"><a href="#%3Cinit%3E(java.lang.String,java.lang.String,java.util.Map)">AllocatedProteinEntry</a></span>&#8203;(java.lang.String&nbsp;name,
java.lang.String&nbsp;pdb,
java.util.Map&lt;java.lang.String,&#8203;java.lang.String&gt;&nbsp;chainSequences)</code></th>
<td class="col-last">
<div class="block">Constructor of <a href="AllocatedProteinEntry.html" title="class in datastructure"><code>AllocatedProteinEntry</code></a></div>
</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ========== METHOD SUMMARY =========== -->
<li>
<section class="method-summary" id="method.summary">
<h2>Method Summary</h2>
<div class="member-summary" id="method-summary-table">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button role="tab" aria-selected="true" aria-controls="method-summary-table.tabpanel" tabindex="0" onkeydown="switchTab(event)" id="t0" class="active-table-tab">All Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t2" class="table-tab" onclick="show(2);">Instance Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t4" class="table-tab" onclick="show(8);">Concrete Methods</button></div>
<div id="method-summary-table.tabpanel" role="tabpanel">
<table class="summary-table" aria-labelledby="t0">
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Method</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color" id="i0">
<td class="col-first"><code>void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#addProteoform(datastructure.Feature,java.lang.String,java.util.concurrent.ConcurrentSkipListMap)">addProteoform</a></span>&#8203;(<a href="FeatureEntry.html" title="class in datastructure">FeatureEntry</a>&nbsp;fEntry,
java.lang.String&nbsp;sId,
java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;java.lang.String&gt;&nbsp;variants)</code></th>
<td class="col-last">
<div class="block">Adds a new <a href="ProteoformEntry.html" title="class in datastructure"><code>ProteoformEntry</code></a> to the <a href="#proteoforms"><code>proteoforms</code></a> structure.</div>
</td>
</tr>
<tr class="row-color" id="i1">
<td class="col-first"><code>private java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#generateProteoformName(java.lang.String)">generateProteoformName</a></span>&#8203;(java.lang.String&nbsp;vSwab)</code></th>
<td class="col-last">
<div class="block">Returns a <code>String</code> that is used as internal proteoform name.</div>
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="inherited-list">
<h3 id="methods.inherited.from.class.java.lang.Object">Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></div>
</section>
</li>
</ul>
</section>
<section class="details">
<ul class="details-list">
<!-- ============ FIELD DETAIL =========== -->
<li>
<section class="field-details" id="field.detail">
<h2>Field Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="name">
<h3>name</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">name</span></div>
<div class="block"><code>String</code>; The internal name of the protein.</div>
</section>
</li>
<li>
<section class="detail" id="pdb">
<h3>pdb</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">pdb</span></div>
<div class="block"><code>String</code> representation of a `pdb` file containing protein structure information of a coding gene feature.</div>
</section>
</li>
<li>
<section class="detail" id="chainSequences">
<h3>chainSequences</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">chainSequences</span></div>
<div class="block">Aligned reference amino acid sequences, one per protein chain, of the entry.</div>
</section>
</li>
<li>
<section class="detail" id="WILD_TYPE_PROTEOFORM_ID">
<h3>WILD_TYPE_PROTEOFORM_ID</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">WILD_TYPE_PROTEOFORM_ID</span></div>
<div class="block"><code>String</code> identifier used to store the wild type proteoform.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#datastructure.AllocatedProteinEntry.WILD_TYPE_PROTEOFORM_ID">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="proteoforms">
<h3>proteoforms</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;<a href="ProteoformEntry.html" title="class in datastructure">ProteoformEntry</a>&gt;</span>&nbsp;<span class="member-name">proteoforms</span></div>
<div class="block">Map structure mapping <code>String</code>s (representing proteoform names) to <a href="ProteoformEntry.html" title="class in datastructure"><code>ProteoformEntry</code></a>s</div>
</section>
</li>
<li>
<section class="detail" id="variants">
<h3>variants</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;<a href="AminoacidVariantAnnotationEntry.html" title="class in datastructure">AminoacidVariantAnnotationEntry</a>&gt;&gt;</span>&nbsp;<span class="member-name">variants</span></div>
<div class="block">Hierarchical map structure that uses <code>String</code>s as keys and point to <a href="AminoacidVariantAnnotationEntry.html" title="class in datastructure"><code>AminoacidVariantAnnotationEntry</code></a>s.
 <p>
 The first and second layer of the map structure use positions wrt. reference feature and the alternate amino-acid content as keys, respectively.</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<li>
<section class="constructor-details" id="constructor.detail">
<h2>Constructor Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="&lt;init&gt;(java.lang.String,java.lang.String,java.util.Map)">
<h3>AllocatedProteinEntry</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="member-name">AllocatedProteinEntry</span>&#8203;(<span class="parameters">java.lang.String&nbsp;name,
java.lang.String&nbsp;pdb,
java.util.Map&lt;java.lang.String,&#8203;java.lang.String&gt;&nbsp;chainSequences)</span></div>
<div class="block">Constructor of <a href="AllocatedProteinEntry.html" title="class in datastructure"><code>AllocatedProteinEntry</code></a></div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>name</code> - <code>String</code>; The internal name of the entry.</dd>
<dd><code>pdb</code> - <code>String</code>; The file content of a .pdb format file.</dd>
<dd><code>chainSequences</code> - <code>HashMap</code>; Mapping chain ids (in the form of <code>String</code>s) to chain amino-acid sequences (in the form of <code>String</code>s).</dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
<!-- ============ METHOD DETAIL ========== -->
<li>
<section class="method-details" id="method.detail">
<h2>Method Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="generateProteoformName(java.lang.String)">
<h3>generateProteoformName</h3>
<div class="member-signature"><span class="modifiers">private</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">generateProteoformName</span>&#8203;(<span class="parameters">java.lang.String&nbsp;vSwab)</span></div>
<div class="block">Returns a <code>String</code> that is used as internal proteoform name.
 <p>
 These names are generated by concatenating the prefix 'PF' with the HashCode of the <a href="ProteoformEntry.html" title="class in datastructure"><code>ProteoformEntry</code></a>
 mandatory 'VSWAB' annotation value converted to a <code>String</code> for which a negative sign is converted to a one
 and which in total was padded with prepending 0s to a length of 10.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>vSwab</code> - The <code>String</code> value of the 'VSWAB' annotation of a <a href="ProteoformEntry.html" title="class in datastructure"><code>ProteoformEntry</code></a>.</dd>
<dt>Returns:</dt>
<dd><code>String</code> intended to be used as internal proteoform name.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="addProteoform(datastructure.Feature,java.lang.String,java.util.concurrent.ConcurrentSkipListMap)">
<h3>addProteoform</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">addProteoform</span>&#8203;(<span class="parameters"><a href="FeatureEntry.html" title="class in datastructure">FeatureEntry</a>&nbsp;fEntry,
java.lang.String&nbsp;sId,
java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;java.lang.String&gt;&nbsp;variants)</span></div>
<div class="block">Adds a new <a href="ProteoformEntry.html" title="class in datastructure"><code>ProteoformEntry</code></a> to the <a href="#proteoforms"><code>proteoforms</code></a> structure.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>fEntry</code> - The <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a> to whose <a href="FeatureEntry.html#allocatedProtein"><code>FeatureEntry.allocatedProtein</code></a> the entry shall be added.</dd>
<dd><code>sId</code> - The <code>String</code> sample identifier of any <a href="SampleEntry.html" title="class in datastructure"><code>SampleEntry</code></a> of which the proteoform was derived.</dd>
<dd><code>variants</code> - <code>ConcurrentSkipListMap</code> of <code>String</code>s that maps positions wrt. the reference wild-type
                 proteoform to alternate amino-acid contents.</dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
</ul>
</section>
<!-- ========= END OF CLASS DATA ========= -->
</main>
<footer role="contentinfo">
<nav role="navigation">
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottom-nav" id="navbar.bottom">
<div class="skip-nav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.bottom.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
</div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<span class="skip-nav" id="skip.navbar.bottom">
<!--   -->
</span></nav>
</footer>
</div>
</div>
</body>
</html>
