<!DOCTYPE HTML>
<!-- NewPage -->
<html lang="de">
<head>
<!-- Generated by javadoc (15) on Mon Jul 25 13:17:44 CEST 2022 -->
<title>FeatureEntry</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="dc.created" content="2022-07-25">
<meta name="description" content="declaration: package: datastructure, class: FeatureEntry">
<meta name="generator" content="javadoc/ClassWriterImpl">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
<script type="text/javascript" src="../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../script-dir/jquery-ui.min.js"></script>
</head>
<body class="class-declaration-page">
<script type="text/javascript">var data = {"i0":9,"i1":9};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],8:["t4","Concrete Methods"]};
var altColor = "alt-color";
var rowColor = "row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar.top">
<div class="skip-nav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.top.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search">SEARCH:</label>
<input type="text" id="search" value="search" disabled="disabled">
<input type="reset" id="reset" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip.navbar.top">
<!--   -->
</span></nav>
</header>
<div class="flex-content">
<main role="main">
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="sub-title"><span class="package-label-in-type">Package</span>&nbsp;<a href="package-summary.html">datastructure</a></div>
<h1 title="Class FeatureEntry" class="title">Class FeatureEntry</h1>
</div>
<div class="inheritance" title="Inheritance Tree">java.lang.Object
<div class="inheritance">datastructure.Feature</div>
</div>
<section class="description">
<hr>
<pre>public final class <span class="type-name-label">FeatureEntry</span>
extends java.lang.Object</pre>
<div class="block">Internal representation of a reference sequence location that is subject to analysis. May represent the full
 genome, a single gene, contigs or plasmids and chromosomes.
 <p>
 Pairs of instances of this class and <a href="SampleEntry.html" title="class in datastructure"><code>SampleEntry</code></a> instances are used internally as so called 'Run
 entries' to specify a set of analysis tasks, i.e. which sample is analyzed with respect to which specified
 reference feature.</div>
<dl class="notes">
<dt>Since:</dt>
<dd>2.0</dd>
<dt>Version:</dt>
<dd>2.1</dd>
<dt>Author:</dt>
<dd>Simon Hackl</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<!-- =========== FIELD SUMMARY =========== -->
<li>
<section class="field-summary" id="field.summary">
<h2>Field Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Fields</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Field</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<td class="col-first"><code><a href="AllocatedProteinEntry.html" title="class in datastructure">AllocatedProteinEntry</a></code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#allocatedProtein">allocatedProtein</a></span></code></th>
<td class="col-last">
<div class="block">A <a href="AllocatedProteinEntry.html" title="class in datastructure"><code>AllocatedProteinEntry</code></a> object containing information about the reference protein structure of the possible
 protein product of the gene represented by this <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a> and all discovered proteoforms.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#annotations">annotations</a></span></code></th>
<td class="col-last">
<div class="block">A <code>HashMap</code> yielding any meta-information <code>String</code> key-value pairs about this <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#chromosome">chromosome</a></span></code></th>
<td class="col-last">
<div class="block">The location of the entry on the reference, i.e. for genes the contig or chromosome the feature is located on.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>int</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#end">end</a></span></code></th>
<td class="col-last">
<div class="block">The 1-based indexed end position of the feature.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>boolean</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#isSense">isSense</a></span></code></th>
<td class="col-last">
<div class="block">Indicates if the feature is located on the sense strand.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#name">name</a></span></code></th>
<td class="col-last">
<div class="block">The internal name of the entry.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#nucleotideSequence">nucleotideSequence</a></span></code></th>
<td class="col-last">
<div class="block">The reference DNA sequence of the entry.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>java.io.File</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#pdbFile">pdbFile</a></span></code></th>
<td class="col-last">
<div class="block">A <code>File</code> object pointing to a .pdb format file containing the protein structure information of
 <a href="#allocatedProtein"><code>allocatedProtein</code></a>.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>int</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#start">start</a></span></code></th>
<td class="col-last">
<div class="block">The 1-based indexed starting position of the feature.</div>
</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<li>
<section class="constructor-summary" id="constructor.summary">
<h2>Constructor Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Constructors</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Constructor</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<th class="col-constructor-name" scope="row"><code><span class="member-name-link"><a href="#%3Cinit%3E(java.lang.String,java.lang.String,int,int)">FeatureEntry</a></span>&#8203;(java.lang.String&nbsp;entryName,
java.lang.String&nbsp;entryLocation,
int&nbsp;entryStart,
int&nbsp;entryEnd)</code></th>
<td class="col-last">
<div class="block">Constructor of <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>.</div>
</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ========== METHOD SUMMARY =========== -->
<li>
<section class="method-summary" id="method.summary">
<h2>Method Summary</h2>
<div class="member-summary" id="method-summary-table">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button role="tab" aria-selected="true" aria-controls="method-summary-table.tabpanel" tabindex="0" onkeydown="switchTab(event)" id="t0" class="active-table-tab">All Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t1" class="table-tab" onclick="show(1);">Static Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t4" class="table-tab" onclick="show(8);">Concrete Methods</button></div>
<div id="method-summary-table.tabpanel" role="tabpanel">
<table class="summary-table" aria-labelledby="t0">
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Method</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color" id="i0">
<td class="col-first"><code>static void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#imputeProtein(datastructure.Feature)">imputeProtein</a></span>&#8203;(<a href="FeatureEntry.html" title="class in datastructure">FeatureEntry</a>&nbsp;feature)</code></th>
<td class="col-last">
<div class="block">Method to add the protein wild-type information to any <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>.</div>
</td>
</tr>
<tr class="row-color" id="i1">
<td class="col-first"><code>static void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#imputeSequence(datastructure.Feature,datastructure.FastaContainer)">imputeSequence</a></span>&#8203;(<a href="FeatureEntry.html" title="class in datastructure">FeatureEntry</a>&nbsp;feature,
<a href="FastaContainer.html" title="class in datastructure">FastaContainer</a>&nbsp;referenceSequence)</code></th>
<td class="col-last">
<div class="block">Method to add the nucleotide sequence to any <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>.</div>
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="inherited-list">
<h3 id="methods.inherited.from.class.java.lang.Object">Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></div>
</section>
</li>
</ul>
</section>
<section class="details">
<ul class="details-list">
<!-- ============ FIELD DETAIL =========== -->
<li>
<section class="field-details" id="field.detail">
<h2>Field Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="name">
<h3>name</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">name</span></div>
<div class="block">The internal name of the entry.</div>
</section>
</li>
<li>
<section class="detail" id="nucleotideSequence">
<h3>nucleotideSequence</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">nucleotideSequence</span></div>
<div class="block">The reference DNA sequence of the entry.</div>
</section>
</li>
<li>
<section class="detail" id="chromosome">
<h3>chromosome</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">chromosome</span></div>
<div class="block">The location of the entry on the reference, i.e. for genes the contig or chromosome the feature is located on.</div>
</section>
</li>
<li>
<section class="detail" id="start">
<h3>start</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="member-name">start</span></div>
<div class="block">The 1-based indexed starting position of the feature.</div>
</section>
</li>
<li>
<section class="detail" id="end">
<h3>end</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="member-name">end</span></div>
<div class="block">The 1-based indexed end position of the feature.</div>
</section>
</li>
<li>
<section class="detail" id="isSense">
<h3>isSense</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">boolean</span>&nbsp;<span class="member-name">isSense</span></div>
<div class="block">Indicates if the feature is located on the sense strand.</div>
</section>
</li>
<li>
<section class="detail" id="allocatedProtein">
<h3>allocatedProtein</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type"><a href="AllocatedProteinEntry.html" title="class in datastructure">AllocatedProteinEntry</a></span>&nbsp;<span class="member-name">allocatedProtein</span></div>
<div class="block">A <a href="AllocatedProteinEntry.html" title="class in datastructure"><code>AllocatedProteinEntry</code></a> object containing information about the reference protein structure of the possible
 protein product of the gene represented by this <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a> and all discovered proteoforms.</div>
</section>
</li>
<li>
<section class="detail" id="pdbFile">
<h3>pdbFile</h3>
<div class="member-signature"><span class="modifiers">public transient</span>&nbsp;<span class="return-type">java.io.File</span>&nbsp;<span class="member-name">pdbFile</span></div>
<div class="block">A <code>File</code> object pointing to a .pdb format file containing the protein structure information of
 <a href="#allocatedProtein"><code>allocatedProtein</code></a>.</div>
</section>
</li>
<li>
<section class="detail" id="annotations">
<h3>annotations</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">annotations</span></div>
<div class="block">A <code>HashMap</code> yielding any meta-information <code>String</code> key-value pairs about this <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>.</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<li>
<section class="constructor-details" id="constructor.detail">
<h2>Constructor Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="&lt;init&gt;(java.lang.String,java.lang.String,int,int)">
<h3>FeatureEntry</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="member-name">FeatureEntry</span>&#8203;(<span class="parameters">java.lang.String&nbsp;entryName,
java.lang.String&nbsp;entryLocation,
int&nbsp;entryStart,
int&nbsp;entryEnd)</span>
             throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Constructor of <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>entryName</code> - <code>String</code> representing the internal name of the reference feature to analyze.</dd>
<dd><code>entryLocation</code> - <code>String</code> the name of the reference location (contig, chromosome, plasmid) the feature
                      is located on.</dd>
<dd><code>entryStart</code> - <code>Integer</code> The 1-based indexed starting position of the feature on the reference.</dd>
<dd><code>entryEnd</code> - <code>Integer</code> The 1-based indexed end position of the feature on the reference.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If the specified locus is ambiguous.</dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
<!-- ============ METHOD DETAIL ========== -->
<li>
<section class="method-details" id="method.detail">
<h2>Method Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="imputeSequence(datastructure.Feature,datastructure.FastaContainer)">
<h3>imputeSequence</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">imputeSequence</span>&#8203;(<span class="parameters"><a href="FeatureEntry.html" title="class in datastructure">FeatureEntry</a>&nbsp;feature,
<a href="FastaContainer.html" title="class in datastructure">FastaContainer</a>&nbsp;referenceSequence)</span></div>
<div class="block">Method to add the nucleotide sequence to any <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>feature</code> - The <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a> to which nucleotide sequence information shall be added.</dd>
<dd><code>referenceSequence</code> - <a href="FastaContainer.html" title="class in datastructure"><code>FastaContainer</code></a> representation of the parent/reference sequence, i.e. reference
                          genome or genome assembly.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="imputeProtein(datastructure.Feature)">
<h3>imputeProtein</h3>
<div class="member-signature"><span class="modifiers">public static</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">imputeProtein</span>&#8203;(<span class="parameters"><a href="FeatureEntry.html" title="class in datastructure">FeatureEntry</a>&nbsp;feature)</span>
                          throws <span class="exceptions">java.io.IOException,
<a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Method to add the protein wild-type information to any <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>.
 <p>
 The specified <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a> has to specify the <a href="#pdbFile"><code>pdbFile</code></a> property.
 <p>
 - The parsed protein amino-acid sequence is compared with the translated amino-acid sequence of the
 <a href="#nucleotideSequence"><code>nucleotideSequence</code></a> property by pairwise sequence alignment. While the translated sequence
 is allowed to be longer than the protein sequence, this is not allowed vice versa.
 - The internally stored <a href="#allocatedProtein"><code>allocatedProtein</code></a>s <a href="AllocatedProteinEntry.html#pdb"><code>AllocatedProteinEntry.pdb</code></a> will be matched
 in terms of residue numbers to the information gained from the alignment, i.e. if the .pdb starts with residue 10,
 but the sequences align at the beginning the first residue number in the internal .pdb is set to be 1; if the
 .pdb starts with residue 1, but the sequence (of the protein) has a gap of length 9 at the beginning in the
 alignment the first residue number in the internal .pdb is set to be 10.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>feature</code> - The <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a> to which protein information shall be added.</dd>
<dt>Throws:</dt>
<dd><code>java.io.IOException</code> - If the .pdb file pointed to by the <a href="#pdbFile"><code>pdbFile</code></a> can not be parsed.</dd>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If either the translated <a href="#nucleotideSequence"><code>nucleotideSequence</code></a> yields internal
                            termination codons or contains any gaps when aligned to the amino-acid sequence derived from the .pdb fiel.</dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
</ul>
</section>
<!-- ========= END OF CLASS DATA ========= -->
</main>
<footer role="contentinfo">
<nav role="navigation">
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottom-nav" id="navbar.bottom">
<div class="skip-nav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.bottom.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
</div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<span class="skip-nav" id="skip.navbar.bottom">
<!--   -->
</span></nav>
</footer>
</div>
</div>
</body>
</html>
