<!DOCTYPE HTML>
<!-- NewPage -->
<html lang="de">
<head>
<!-- Generated by javadoc (15) on Mon Jul 25 13:17:44 CEST 2022 -->
<title>VariantsDictionary</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta name="dc.created" content="2022-07-25">
<meta name="description" content="declaration: package: datastructure, class: VariantsDictionary">
<meta name="generator" content="javadoc/ClassWriterImpl">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
<script type="text/javascript" src="../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../script-dir/jquery-ui.min.js"></script>
</head>
<body class="class-declaration-page">
<script type="text/javascript">var data = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10,"i5":10,"i6":10};
var tabs = {65535:["t0","All Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "alt-color";
var rowColor = "row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar.top">
<div class="skip-nav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.top.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search">SEARCH:</label>
<input type="text" id="search" value="search" disabled="disabled">
<input type="reset" id="reset" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip.navbar.top">
<!--   -->
</span></nav>
</header>
<div class="flex-content">
<main role="main">
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="sub-title"><span class="package-label-in-type">Package</span>&nbsp;<a href="package-summary.html">datastructure</a></div>
<h1 title="Class VariantsDictionary" class="title">Class VariantsDictionary</h1>
</div>
<div class="inheritance" title="Inheritance Tree">java.lang.Object
<div class="inheritance">datastructure.MusialStorage</div>
</div>
<section class="description">
<hr>
<pre>public class <span class="type-name-label">VariantsDictionary</span>
extends java.lang.Object</pre>
<div class="block">Main data structure to store information about variants.</div>
<dl class="notes">
<dt>Since:</dt>
<dd>2.1</dd>
<dt>Version:</dt>
<dd>2.1</dd>
<dt>Author:</dt>
<dd>Simon Hackl</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<!-- =========== FIELD SUMMARY =========== -->
<li>
<section class="field-summary" id="field.summary">
<h2>Field Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Fields</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Field</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#ATTRIBUTE_VARIANT_SWAB_NAME">ATTRIBUTE_VARIANT_SWAB_NAME</a></span></code></th>
<td class="col-last">
<div class="block">Static property to use as suffix for the 'VSWAB' annotation key in <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>s and <a href="SampleEntry.html" title="class in datastructure"><code>SampleEntry</code></a>s.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#chromosome">chromosome</a></span></code></th>
<td class="col-last">
<div class="block">The chromosome name on which all maintained <a href="#features"><code>features</code></a> are located.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#date">date</a></span></code></th>
<td class="col-last">
<div class="block">Date tag for time stamp generation.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>java.util.HashMap&lt;java.lang.String,&#8203;<a href="FeatureEntry.html" title="class in datastructure">FeatureEntry</a>&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#features">features</a></span></code></th>
<td class="col-last">
<div class="block">Map of <code>String</code> (key) / <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a> (value) pairs; All (gene) features that are maintained.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>java.util.concurrent.ConcurrentSkipListSet&lt;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#novelVariants">novelVariants</a></span></code></th>
<td class="col-last">
<div class="block">Transient property to store novel variants during the execution of the updateVDict module/task.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code><a href="VariantsDictionaryParameters.html" title="class in datastructure">VariantsDictionaryParameters</a></code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#parameters">parameters</a></span></code></th>
<td class="col-last">
<div class="block">Set of parameters, a <a href="VariantsDictionaryParameters.html" title="class in datastructure"><code>VariantsDictionaryParameters</code></a> instance, used for variantInformation filtering.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>java.util.HashMap&lt;java.lang.String,&#8203;<a href="SampleEntry.html" title="class in datastructure">SampleEntry</a>&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#samples">samples</a></span></code></th>
<td class="col-last">
<div class="block">Map of <code>String</code> (key) / <a href="SampleEntry.html" title="class in datastructure"><code>SampleEntry</code></a> (value) pairs; All samples that are maintained.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#software">software</a></span></code></th>
<td class="col-last">
<div class="block">Software name for software version meta-information generation.</div>
</td>
</tr>
<tr class="alt-color">
<td class="col-first"><code>java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.Integer,&#8203;java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;<a href="NucleotideVariantAnnotationEntry.html" title="class in datastructure">NucleotideVariantAnnotationEntry</a>&gt;&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#variants">variants</a></span></code></th>
<td class="col-last">
<div class="block">Hierarchical map structure to store variants wrt. maintained features and samples.</div>
</td>
</tr>
<tr class="row-color">
<td class="col-first"><code>static java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#WILD_TYPE_SAMPLE_ID">WILD_TYPE_SAMPLE_ID</a></span></code></th>
<td class="col-last">
<div class="block">Static property to use as the sample name/id for the wild type.</div>
</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<li>
<section class="constructor-summary" id="constructor.summary">
<h2>Constructor Summary</h2>
<div class="member-summary">
<table class="summary-table">
<caption><span>Constructors</span></caption>
<thead>
<tr>
<th class="col-first" scope="col">Constructor</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color">
<th class="col-constructor-name" scope="row"><code><span class="member-name-link"><a href="#%3Cinit%3E(java.lang.Double,java.lang.Double,java.lang.Double,java.lang.Double,java.lang.Double,java.lang.String)">VariantsDictionary</a></span>&#8203;(java.lang.Double&nbsp;minCoverage,
java.lang.Double&nbsp;minFrequency,
java.lang.Double&nbsp;minHet,
java.lang.Double&nbsp;maxHet,
java.lang.Double&nbsp;minQuality,
java.lang.String&nbsp;chromosome)</code></th>
<td class="col-last">
<div class="block">Constructor of <a href="VariantsDictionary.html" title="class in datastructure"><code>VariantsDictionary</code></a>.</div>
</td>
</tr>
</tbody>
</table>
</div>
</section>
</li>
<!-- ========== METHOD SUMMARY =========== -->
<li>
<section class="method-summary" id="method.summary">
<h2>Method Summary</h2>
<div class="member-summary" id="method-summary-table">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button role="tab" aria-selected="true" aria-controls="method-summary-table.tabpanel" tabindex="0" onkeydown="switchTab(event)" id="t0" class="active-table-tab">All Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t2" class="table-tab" onclick="show(2);">Instance Methods</button><button role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" id="t4" class="table-tab" onclick="show(8);">Concrete Methods</button></div>
<div id="method-summary-table.tabpanel" role="tabpanel">
<table class="summary-table" aria-labelledby="t0">
<thead>
<tr>
<th class="col-first" scope="col">Modifier and Type</th>
<th class="col-second" scope="col">Method</th>
<th class="col-last" scope="col">Description</th>
</tr>
</thead>
<tbody>
<tr class="alt-color" id="i0">
<td class="col-first"><code>void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#addVariant(java.lang.String,int,java.lang.String,java.lang.String,java.lang.String,boolean,boolean,double,double,double)">addVariant</a></span>&#8203;(java.lang.String&nbsp;featureId,
int&nbsp;referencePosition,
java.lang.String&nbsp;variantContent,
java.lang.String&nbsp;referenceContent,
java.lang.String&nbsp;sampleId,
boolean&nbsp;isPrimary,
boolean&nbsp;isRejected,
double&nbsp;quality,
double&nbsp;coverage,
double&nbsp;frequency)</code></th>
<td class="col-last">
<div class="block">Adds information about a variantInformation to <a href="#variants"><code>variants</code></a>.</div>
</td>
</tr>
<tr class="row-color" id="i1">
<td class="col-first"><code>void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#addVariantAnnotation(int,java.lang.String,java.util.Map)">addVariantAnnotation</a></span>&#8203;(int&nbsp;position,
java.lang.String&nbsp;content,
java.util.Map&lt;java.lang.String,&#8203;java.lang.String&gt;&nbsp;annotations)</code></th>
<td class="col-last">
<div class="block">Adds annotations to an existing variantInformation.</div>
</td>
</tr>
<tr class="alt-color" id="i2">
<td class="col-first"><code>void</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#dump(java.io.File)">dump</a></span>&#8203;(java.io.File&nbsp;outfile)</code></th>
<td class="col-last">
<div class="block">Dumps this <a href="VariantsDictionary.html" title="class in datastructure"><code>VariantsDictionary</code></a> instance into a JSON format file.</div>
</td>
</tr>
<tr class="row-color" id="i3">
<td class="col-first"><code>java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getNucleotideSequence(java.lang.String,java.lang.String)">getNucleotideSequence</a></span>&#8203;(java.lang.String&nbsp;fId,
java.lang.String&nbsp;sId)</code></th>
<td class="col-last">
<div class="block">Extracts a <code>String</code> yielding the nucleotide sequence with all variants of one sample being incorporated for one feature.</div>
</td>
</tr>
<tr class="alt-color" id="i4">
<td class="col-first"><code>java.lang.String</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getProteoformSequence(java.lang.String,java.lang.String)">getProteoformSequence</a></span>&#8203;(java.lang.String&nbsp;fId,
java.lang.String&nbsp;pfId)</code></th>
<td class="col-last">
<div class="block">Extracts a <code>String</code> yielding the amino-acid sequence with all variants of one proteoform being incorporated
 for one feature.</div>
</td>
</tr>
<tr class="row-color" id="i5">
<td class="col-first"><code>java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getProteoformVariants(java.lang.String,java.lang.String,boolean)">getProteoformVariants</a></span>&#8203;(java.lang.String&nbsp;fId,
java.lang.String&nbsp;pfId,
boolean&nbsp;includeDeletions)</code></th>
<td class="col-last">
<div class="block">Extracts a <code>HashMap</code> of <code>String</code>/<code>String</code> key/value pairs reflecting all (filtered) variants
 for one proteoform and feature.</div>
</td>
</tr>
<tr class="alt-color" id="i6">
<td class="col-first"><code>java.util.HashMap&lt;java.lang.Integer,&#8203;java.lang.String&gt;</code></td>
<th class="col-second" scope="row"><code><span class="member-name-link"><a href="#getSampleVariants(java.lang.String,java.lang.String)">getSampleVariants</a></span>&#8203;(java.lang.String&nbsp;fId,
java.lang.String&nbsp;sId)</code></th>
<td class="col-last">
<div class="block">Extracts a <code>HashMap</code> of <code>Integer</code>/<code>String</code> key/value pairs reflecting all (filtered) variants
 for one sample and feature.</div>
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="inherited-list">
<h3 id="methods.inherited.from.class.java.lang.Object">Methods inherited from class&nbsp;java.lang.Object</h3>
<code>clone, equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></div>
</section>
</li>
</ul>
</section>
<section class="details">
<ul class="details-list">
<!-- ============ FIELD DETAIL =========== -->
<li>
<section class="field-details" id="field.detail">
<h2>Field Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="parameters">
<h3>parameters</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type"><a href="VariantsDictionaryParameters.html" title="class in datastructure">VariantsDictionaryParameters</a></span>&nbsp;<span class="member-name">parameters</span></div>
<div class="block">Set of parameters, a <a href="VariantsDictionaryParameters.html" title="class in datastructure"><code>VariantsDictionaryParameters</code></a> instance, used for variantInformation filtering.</div>
</section>
</li>
<li>
<section class="detail" id="features">
<h3>features</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;<a href="FeatureEntry.html" title="class in datastructure">FeatureEntry</a>&gt;</span>&nbsp;<span class="member-name">features</span></div>
<div class="block">Map of <code>String</code> (key) / <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a> (value) pairs; All (gene) features that are maintained.</div>
</section>
</li>
<li>
<section class="detail" id="samples">
<h3>samples</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;<a href="SampleEntry.html" title="class in datastructure">SampleEntry</a>&gt;</span>&nbsp;<span class="member-name">samples</span></div>
<div class="block">Map of <code>String</code> (key) / <a href="SampleEntry.html" title="class in datastructure"><code>SampleEntry</code></a> (value) pairs; All samples that are maintained.</div>
</section>
</li>
<li>
<section class="detail" id="software">
<h3>software</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">software</span></div>
<div class="block">Software name for software version meta-information generation.</div>
</section>
</li>
<li>
<section class="detail" id="date">
<h3>date</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">date</span></div>
<div class="block">Date tag for time stamp generation.</div>
</section>
</li>
<li>
<section class="detail" id="chromosome">
<h3>chromosome</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">chromosome</span></div>
<div class="block">The chromosome name on which all maintained <a href="#features"><code>features</code></a> are located.</div>
</section>
</li>
<li>
<section class="detail" id="variants">
<h3>variants</h3>
<div class="member-signature"><span class="modifiers">public final</span>&nbsp;<span class="return-type">java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.Integer,&#8203;java.util.concurrent.ConcurrentSkipListMap&lt;java.lang.String,&#8203;<a href="NucleotideVariantAnnotationEntry.html" title="class in datastructure">NucleotideVariantAnnotationEntry</a>&gt;&gt;</span>&nbsp;<span class="member-name">variants</span></div>
<div class="block">Hierarchical map structure to store variants wrt. maintained features and samples. The first layer represents the
 position on the chromosome. The second layer represents the variantInformation content.</div>
</section>
</li>
<li>
<section class="detail" id="WILD_TYPE_SAMPLE_ID">
<h3>WILD_TYPE_SAMPLE_ID</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">WILD_TYPE_SAMPLE_ID</span></div>
<div class="block">Static property to use as the sample name/id for the wild type.
 TODO: Move to separate class with other attribute names.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#datastructure.MusialStorage.WILD_TYPE_SAMPLE_ID">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="ATTRIBUTE_VARIANT_SWAB_NAME">
<h3>ATTRIBUTE_VARIANT_SWAB_NAME</h3>
<div class="member-signature"><span class="modifiers">public static final</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">ATTRIBUTE_VARIANT_SWAB_NAME</span></div>
<div class="block">Static property to use as suffix for the 'VSWAB' annotation key in <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>s and <a href="SampleEntry.html" title="class in datastructure"><code>SampleEntry</code></a>s.</div>
<dl class="notes">
<dt>See Also:</dt>
<dd><a href="../constant-values.html#datastructure.MusialStorage.ATTRIBUTE_VARIANT_SWAB_NAME">Constant Field Values</a></dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="novelVariants">
<h3>novelVariants</h3>
<div class="member-signature"><span class="modifiers">public transient</span>&nbsp;<span class="return-type">java.util.concurrent.ConcurrentSkipListSet&lt;java.lang.String&gt;</span>&nbsp;<span class="member-name">novelVariants</span></div>
<div class="block">Transient property to store novel variants during the execution of the updateVDict module/task.</div>
</section>
</li>
</ul>
</section>
</li>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<li>
<section class="constructor-details" id="constructor.detail">
<h2>Constructor Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="&lt;init&gt;(java.lang.Double,java.lang.Double,java.lang.Double,java.lang.Double,java.lang.Double,java.lang.String)">
<h3>VariantsDictionary</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="member-name">VariantsDictionary</span>&#8203;(<span class="parameters">java.lang.Double&nbsp;minCoverage,
java.lang.Double&nbsp;minFrequency,
java.lang.Double&nbsp;minHet,
java.lang.Double&nbsp;maxHet,
java.lang.Double&nbsp;minQuality,
java.lang.String&nbsp;chromosome)</span></div>
<div class="block">Constructor of <a href="VariantsDictionary.html" title="class in datastructure"><code>VariantsDictionary</code></a>.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>minCoverage</code> - <code>Double</code>; Minimal read coverage to use for variantInformation filtering.</dd>
<dd><code>minFrequency</code> - <code>Double</code>; Minimal form frequency to use for hom. variantInformation filtering.</dd>
<dd><code>minHet</code> - <code>Double</code>; Minimal form frequency to use for het. variantInformation filtering.</dd>
<dd><code>maxHet</code> - <code>Double</code>; Maximal form frequency to use for het. variantInformation filtering.</dd>
<dd><code>minQuality</code> - <code>Double</code>; Minimal Phred scaled genotyping call quality to use for variantInformation filtering.</dd>
<dd><code>chromosome</code> - <code>String</code>; Name of the chromosome on which maintained <a href="FeatureEntry.html" title="class in datastructure"><code>FeatureEntry</code></a>s are located;
                     Has to reflect the value in the used input .vcf, .fasta and .gff files.</dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
<!-- ============ METHOD DETAIL ========== -->
<li>
<section class="method-details" id="method.detail">
<h2>Method Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="addVariant(java.lang.String,int,java.lang.String,java.lang.String,java.lang.String,boolean,boolean,double,double,double)">
<h3>addVariant</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">addVariant</span>&#8203;(<span class="parameters">java.lang.String&nbsp;featureId,
int&nbsp;referencePosition,
java.lang.String&nbsp;variantContent,
java.lang.String&nbsp;referenceContent,
java.lang.String&nbsp;sampleId,
boolean&nbsp;isPrimary,
boolean&nbsp;isRejected,
double&nbsp;quality,
double&nbsp;coverage,
double&nbsp;frequency)</span></div>
<div class="block">Adds information about a variantInformation to <a href="#variants"><code>variants</code></a>.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>featureId</code> - <code>String</code>; The internal name of the feature for which this variantInformation was detected.</dd>
<dd><code>referencePosition</code> - <code>Integer</code>; The position on the reference chromosome (1-based indexing).</dd>
<dd><code>variantContent</code> - <code>String</code>; The alternate content of the variantInformation; Single letter contents reflect
                          SNVs, multi letter contents reflect SVs.</dd>
<dd><code>referenceContent</code> - <code>String</code>; The reference content of the variantInformation; The length in relation to the
                          variantContent parameter indicates whether a SV is a insertion or deletion.</dd>
<dd><code>sampleId</code> - <code>String</code>; The internal name of the sample of which the variantInformation was called.</dd>
<dd><code>isPrimary</code> - <code>Boolean</code>; Whether the variantInformation call is a primary call, i.e. if it is the variantInformation
                          with the highest frequency.</dd>
<dd><code>isRejected</code> - <code>Boolean</code>; Whether the variantInformation was rejected, i.e. if it failed any filtering
                          criteria.</dd>
<dd><code>quality</code> - <code>Double</code>; The Phred scaled GT quality of the call.</dd>
<dd><code>coverage</code> - <code>Double</code>; The depth of coverage of the call.</dd>
<dd><code>frequency</code> - <code>Double</code>; The allelic frequency of the call.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="addVariantAnnotation(int,java.lang.String,java.util.Map)">
<h3>addVariantAnnotation</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">addVariantAnnotation</span>&#8203;(<span class="parameters">int&nbsp;position,
java.lang.String&nbsp;content,
java.util.Map&lt;java.lang.String,&#8203;java.lang.String&gt;&nbsp;annotations)</span></div>
<div class="block">Adds annotations to an existing variantInformation.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>position</code> - <code>Integer</code>; The position at which the variantInformation is stored.</dd>
<dd><code>content</code> - <code>String</code>; The content of the variantInformation.</dd>
<dd><code>annotations</code> - <code>HashMap</code> of <code>String</code> key/value pairs which reflect the annotations that shall be
                    added.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="dump(java.io.File)">
<h3>dump</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">void</span>&nbsp;<span class="member-name">dump</span>&#8203;(<span class="parameters">java.io.File&nbsp;outfile)</span>
          throws <span class="exceptions"><a href="../exceptions/MusialIOException.html" title="class in exceptions">MusialIOException</a></span></div>
<div class="block">Dumps this <a href="VariantsDictionary.html" title="class in datastructure"><code>VariantsDictionary</code></a> instance into a JSON format file. The output is compressed with gzip
 dependent on the value of <a href="../main/Musial.html#COMPRESS"><code>Musial.COMPRESS</code></a>.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>outfile</code> - <code>File</code> object pointing to the file the VDict should be written to.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialIOException.html" title="class in exceptions">MusialIOException</a></code> - If the output generation fails.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getSampleVariants(java.lang.String,java.lang.String)">
<h3>getSampleVariants</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.Integer,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">getSampleVariants</span>&#8203;(<span class="parameters">java.lang.String&nbsp;fId,
java.lang.String&nbsp;sId)</span></div>
<div class="block">Extracts a <code>HashMap</code> of <code>Integer</code>/<code>String</code> key/value pairs reflecting all (filtered) variants
 for one sample and feature.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>fId</code> - <code>String</code>; The name/id of the feature for which variants shall be extracted.</dd>
<dd><code>sId</code> - <code>String</code>; The name/id of the sample for which variants shall be extracted.</dd>
<dt>Returns:</dt>
<dd><code>HashMap</code> of variants extracted for the specified feature and sample.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getNucleotideSequence(java.lang.String,java.lang.String)">
<h3>getNucleotideSequence</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">getNucleotideSequence</span>&#8203;(<span class="parameters">java.lang.String&nbsp;fId,
java.lang.String&nbsp;sId)</span></div>
<div class="block">Extracts a <code>String</code> yielding the nucleotide sequence with all variants of one sample being incorporated for one feature.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>fId</code> - <code>String</code>; The name/id of the feature for which the sequence shall be extracted.</dd>
<dd><code>sId</code> - <code>String</code>; The name/id of the sample for which the sequence shall be extracted.</dd>
<dt>Returns:</dt>
<dd><code>String</code>; Nucleotide sequence of one sample wrt. one feature.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getProteoformVariants(java.lang.String,java.lang.String,boolean)">
<h3>getProteoformVariants</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">java.util.HashMap&lt;java.lang.String,&#8203;java.lang.String&gt;</span>&nbsp;<span class="member-name">getProteoformVariants</span>&#8203;(<span class="parameters">java.lang.String&nbsp;fId,
java.lang.String&nbsp;pfId,
boolean&nbsp;includeDeletions)</span></div>
<div class="block">Extracts a <code>HashMap</code> of <code>String</code>/<code>String</code> key/value pairs reflecting all (filtered) variants
 for one proteoform and feature.
 <p>
 - Returns null if no protein is allocated to the feature.
 - Keys of the returned map are formatted as X+Y; X is the position wrt. the reference protein and Y the number of
 inserted positions.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>fId</code> - <code>String</code>; The name/id of the feature for which variants shall be extracted.</dd>
<dd><code>pfId</code> - <code>String</code>; The name/id of the proteoform for which variants shall be extracted.</dd>
<dd><code>includeDeletions</code> - <code>Boolean</code> whether deletions shall be included as variants.</dd>
<dt>Returns:</dt>
<dd><code>HashMap</code> of variants extracted for the specified feature and proteoform.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="getProteoformSequence(java.lang.String,java.lang.String)">
<h3>getProteoformSequence</h3>
<div class="member-signature"><span class="modifiers">public</span>&nbsp;<span class="return-type">java.lang.String</span>&nbsp;<span class="member-name">getProteoformSequence</span>&#8203;(<span class="parameters">java.lang.String&nbsp;fId,
java.lang.String&nbsp;pfId)</span>
                                       throws <span class="exceptions"><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></span></div>
<div class="block">Extracts a <code>String</code> yielding the amino-acid sequence with all variants of one proteoform being incorporated
 for one feature.
 <p>
 The translated reference feature sequence will be used as the reference amino-acid sequence.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>fId</code> - <code>String</code>; The name/id of the feature for which the sequence shall be extracted.</dd>
<dd><code>pfId</code> - <code>String</code>; The name/id of the proteoform for which the sequence shall be extracted.</dd>
<dt>Returns:</dt>
<dd><code>String</code>; Amino-acid sequence of one proteoform wrt. one feature.</dd>
<dt>Throws:</dt>
<dd><code><a href="../exceptions/MusialBioException.html" title="class in exceptions">MusialBioException</a></code> - If the translation of the reference sequence fails.</dd>
</dl>
</section>
</li>
</ul>
</section>
</li>
</ul>
</section>
<!-- ========= END OF CLASS DATA ========= -->
</main>
<footer role="contentinfo">
<nav role="navigation">
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottom-nav" id="navbar.bottom">
<div class="skip-nav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar.bottom.firstrow" class="nav-list" title="Navigation">
<li><a href="../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
</div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<span class="skip-nav" id="skip.navbar.bottom">
<!--   -->
</span></nav>
</footer>
</div>
</div>
</body>
</html>
